<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8">
	<title>Daniel Standage</title>
	<meta name="description" content="">
	<meta name="author" content="Daniel S. Standage">

	<!-- HTML5 shim, for IE6-8 support of HTML elements -->
	<!--[if lt IE 9]>
		<script src="https://standage.github.io/theme/html5.js"></script>
	<![endif]-->

	<!-- Styles -->
	<link href="https://standage.github.io/theme/bootstrap.min.css" rel="stylesheet">
	<link href="https://standage.github.io/theme/local.css" rel="stylesheet">
	<link href="https://standage.github.io/theme/pygments.css" rel="stylesheet">

	<!-- Feeds -->
	<link href="https://standage.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Daniel Standage Atom Feed" />


	<link href="https://standage.github.io/feed.xml" type="application/rss+xml" rel="alternate" title="Daniel Standage RSS Feed" />


<script type="text/javascript">
	var disqus_identifier = "how-to-distinguish-perfectly-mapped-reads-from-a-sambam-file.html";
	(function() {
		var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
		dsq.src = 'https://standage-github-io.disqus.com/embed.js';
		(document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
	})();
</script>
</head>
<body>
	<div class="topbar">
	  <div class="topbar-inner">
		<div class="container-fluid">
		  <a class="brand" href="https://standage.github.io/">Daniel Standage</a>
			<ul class="nav">
						<li><a href="https://standage.github.io/pages/cv.html">CV</a></li>
						<li><a href="https://standage.github.io/pages/teaching.html">Teaching</a></li>
						<li><a href="https://standage.github.io/pages/software.html">Software</a></li>
				<li><a href="https://standage.github.io/notebook.html">Notebook</a></li>
			</ul>
			<p class="pull-right"><a href="https://standage.github.io/archives.html">[archives]</a> <a href="https://standage.github.io/categories.html">[categories]</a> <a href="https://standage.github.io/tags.html">[tags]</a></p>
		</div>
	  </div>
	</div>

	<div class="container-fluid">
	  <div class="sidebar">
		<div class="well">
			<img src="https://www.gravatar.com/avatar/55b2cca3b7a11df3a961d57a68705503?size=180" alt="Daniel Standage" />
			<h3>Daniel S. Standage</h3>
			<p>
				253 CCAH<br />
				University of California, Davis<br />
				1 Shields Ave<br />
				Davis, CA 95616<br />
			</p>
			<h3>Affiliations</h3>
			<ul>
				<li><a href="http://ivory.idyll.org/lab/">Lab for Data Intensive Biology</a></li>
				<li><a href="http://brendelgroup.org/">Brendel Group</a></li>
				<li><a href="http://www.public.iastate.edu/~amytoth/Toth_lab/Home.html">Toth Lab</a></li>
			</ul>
			<div class="social">
			<h3>Social</h3>
			<ul>
				<li><a href="http://github.com/standage">GitHub</a></li>
				<li><a href="https://twitter.com/byuhobbes">Twitter</a></li>
				<li><a href="http://stackexchange.com/users/208980/daniel-standage?tab=accounts">StackExchange</a></li>
			</ul>
			</div>
		</div>
	  </div>
	  <div class="content">
	<div class='article'>
		<div class="page-header"><h1>How to distinguish perfectly mapped reads from a SAM/BAM file</h1></div>
		<div class="well small">Permalink: <a class="more" href="https://standage.github.io/how-to-distinguish-perfectly-mapped-reads-from-a-sambam-file.html">2017-04-14</a>
by <a class="url fn" href="https://standage.github.io/author/daniel-s-standage.html">Daniel S. Standage </a>
 in <a href="https://standage.github.io/category/notebook.html">notebook</a>
tags: <a href="https://standage.github.io/tag/bash.html">bash</a> <a href="https://standage.github.io/tag/sam.html">sam</a> <a href="https://standage.github.io/tag/bam.html">bam</a> <a href="https://standage.github.io/tag/mapping.html">mapping</a> </div>
		<div><style type="text/css">/*!
*
* IPython notebook
*
*/
/* CSS font colors for translated ANSI colors. */
.ansibold {
  font-weight: bold;
}
/* use dark versions for foreground, to improve visibility */
.ansiblack {
  color: black;
}
.ansired {
  color: darkred;
}
.ansigreen {
  color: darkgreen;
}
.ansiyellow {
  color: #c4a000;
}
.ansiblue {
  color: darkblue;
}
.ansipurple {
  color: darkviolet;
}
.ansicyan {
  color: steelblue;
}
.ansigray {
  color: gray;
}
/* and light for background, for the same reason */
.ansibgblack {
  background-color: black;
}
.ansibgred {
  background-color: red;
}
.ansibggreen {
  background-color: green;
}
.ansibgyellow {
  background-color: yellow;
}
.ansibgblue {
  background-color: blue;
}
.ansibgpurple {
  background-color: magenta;
}
.ansibgcyan {
  background-color: cyan;
}
.ansibggray {
  background-color: gray;
}
div.cell {
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: vertical;
  -moz-box-align: stretch;
  display: box;
  box-orient: vertical;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: column;
  align-items: stretch;
  border-radius: 2px;
  box-sizing: border-box;
  -moz-box-sizing: border-box;
  -webkit-box-sizing: border-box;
  border-width: 1px;
  border-style: solid;
  border-color: transparent;
  width: 100%;
  padding: 5px;
  /* This acts as a spacer between cells, that is outside the border */
  margin: 0px;
  outline: none;
  border-left-width: 1px;
  padding-left: 5px;
  background: linear-gradient(to right, transparent -40px, transparent 1px, transparent 1px, transparent 100%);
}
div.cell.jupyter-soft-selected {
  border-left-color: #90CAF9;
  border-left-color: #E3F2FD;
  border-left-width: 1px;
  padding-left: 5px;
  border-right-color: #E3F2FD;
  border-right-width: 1px;
  background: #E3F2FD;
}
@media print {
  div.cell.jupyter-soft-selected {
    border-color: transparent;
  }
}
div.cell.selected {
  border-color: #ababab;
  border-left-width: 0px;
  padding-left: 6px;
  background: linear-gradient(to right, #42A5F5 -40px, #42A5F5 5px, transparent 5px, transparent 100%);
}
@media print {
  div.cell.selected {
    border-color: transparent;
  }
}
div.cell.selected.jupyter-soft-selected {
  border-left-width: 0;
  padding-left: 6px;
  background: linear-gradient(to right, #42A5F5 -40px, #42A5F5 7px, #E3F2FD 7px, #E3F2FD 100%);
}
.edit_mode div.cell.selected {
  border-color: #66BB6A;
  border-left-width: 0px;
  padding-left: 6px;
  background: linear-gradient(to right, #66BB6A -40px, #66BB6A 5px, transparent 5px, transparent 100%);
}
@media print {
  .edit_mode div.cell.selected {
    border-color: transparent;
  }
}
.prompt {
  /* This needs to be wide enough for 3 digit prompt numbers: In[100]: */
  min-width: 14ex;
  /* This padding is tuned to match the padding on the CodeMirror editor. */
  padding: 0.4em;
  margin: 0px;
  font-family: monospace;
  text-align: right;
  /* This has to match that of the the CodeMirror class line-height below */
  line-height: 1.21429em;
  /* Don't highlight prompt number selection */
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  /* Use default cursor */
  cursor: default;
}
@media (max-width: 540px) {
  .prompt {
    text-align: left;
  }
}
div.inner_cell {
  min-width: 0;
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: vertical;
  -moz-box-align: stretch;
  display: box;
  box-orient: vertical;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: column;
  align-items: stretch;
  /* Old browsers */
  -webkit-box-flex: 1;
  -moz-box-flex: 1;
  box-flex: 1;
  /* Modern browsers */
  flex: 1;
}
/* input_area and input_prompt must match in top border and margin for alignment */
div.input_area {
  border: 1px solid #cfcfcf;
  border-radius: 2px;
  background: #f7f7f7;
  line-height: 1.21429em;
}
/* This is needed so that empty prompt areas can collapse to zero height when there
   is no content in the output_subarea and the prompt. The main purpose of this is
   to make sure that empty JavaScript output_subareas have no height. */
div.prompt:empty {
  padding-top: 0;
  padding-bottom: 0;
}
div.unrecognized_cell {
  padding: 5px 5px 5px 0px;
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: horizontal;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: horizontal;
  -moz-box-align: stretch;
  display: box;
  box-orient: horizontal;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: row;
  align-items: stretch;
}
div.unrecognized_cell .inner_cell {
  border-radius: 2px;
  padding: 5px;
  font-weight: bold;
  color: red;
  border: 1px solid #cfcfcf;
  background: #eaeaea;
}
div.unrecognized_cell .inner_cell a {
  color: inherit;
  text-decoration: none;
}
div.unrecognized_cell .inner_cell a:hover {
  color: inherit;
  text-decoration: none;
}
@media (max-width: 540px) {
  div.unrecognized_cell > div.prompt {
    display: none;
  }
}
div.code_cell {
  /* avoid page breaking on code cells when printing */
}
@media print {
  div.code_cell {
    page-break-inside: avoid;
  }
}
/* any special styling for code cells that are currently running goes here */
div.input {
  page-break-inside: avoid;
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: horizontal;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: horizontal;
  -moz-box-align: stretch;
  display: box;
  box-orient: horizontal;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: row;
  align-items: stretch;
}
@media (max-width: 540px) {
  div.input {
    /* Old browsers */
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-box-align: stretch;
    display: -moz-box;
    -moz-box-orient: vertical;
    -moz-box-align: stretch;
    display: box;
    box-orient: vertical;
    box-align: stretch;
    /* Modern browsers */
    display: flex;
    flex-direction: column;
    align-items: stretch;
  }
}
/* input_area and input_prompt must match in top border and margin for alignment */
div.input_prompt {
  color: #303F9F;
  border-top: 1px solid transparent;
}
div.input_area > div.highlight {
  margin: 0.4em;
  border: none;
  padding: 0px;
  background-color: transparent;
}
div.input_area > div.highlight > pre {
  margin: 0px;
  border: none;
  padding: 0px;
  background-color: transparent;
}
/* The following gets added to the <head> if it is detected that the user has a
 * monospace font with inconsistent normal/bold/italic height.  See
 * notebookmain.js.  Such fonts will have keywords vertically offset with
 * respect to the rest of the text.  The user should select a better font.
 * See: https://github.com/ipython/ipython/issues/1503
 *
 * .CodeMirror span {
 *      vertical-align: bottom;
 * }
 */
.CodeMirror {
  line-height: 1.21429em;
  /* Changed from 1em to our global default */
  font-size: 14px;
  height: auto;
  /* Changed to auto to autogrow */
  background: none;
  /* Changed from white to allow our bg to show through */
}
.CodeMirror-scroll {
  /*  The CodeMirror docs are a bit fuzzy on if overflow-y should be hidden or visible.*/
  /*  We have found that if it is visible, vertical scrollbars appear with font size changes.*/
  overflow-y: hidden;
  overflow-x: auto;
}
.CodeMirror-lines {
  /* In CM2, this used to be 0.4em, but in CM3 it went to 4px. We need the em value because */
  /* we have set a different line-height and want this to scale with that. */
  padding: 0.4em;
}
.CodeMirror-linenumber {
  padding: 0 8px 0 4px;
}
.CodeMirror-gutters {
  border-bottom-left-radius: 2px;
  border-top-left-radius: 2px;
}
.CodeMirror pre {
  /* In CM3 this went to 4px from 0 in CM2. We need the 0 value because of how we size */
  /* .CodeMirror-lines */
  padding: 0;
  border: 0;
  border-radius: 0;
}
/*

Original style from softwaremaniacs.org (c) Ivan Sagalaev <Maniac@SoftwareManiacs.Org>
Adapted from GitHub theme

*/
.highlight-base {
  color: #000;
}
.highlight-variable {
  color: #000;
}
.highlight-variable-2 {
  color: #1a1a1a;
}
.highlight-variable-3 {
  color: #333333;
}
.highlight-string {
  color: #BA2121;
}
.highlight-comment {
  color: #408080;
  font-style: italic;
}
.highlight-number {
  color: #080;
}
.highlight-atom {
  color: #88F;
}
.highlight-keyword {
  color: #008000;
  font-weight: bold;
}
.highlight-builtin {
  color: #008000;
}
.highlight-error {
  color: #f00;
}
.highlight-operator {
  color: #AA22FF;
  font-weight: bold;
}
.highlight-meta {
  color: #AA22FF;
}
/* previously not defined, copying from default codemirror */
.highlight-def {
  color: #00f;
}
.highlight-string-2 {
  color: #f50;
}
.highlight-qualifier {
  color: #555;
}
.highlight-bracket {
  color: #997;
}
.highlight-tag {
  color: #170;
}
.highlight-attribute {
  color: #00c;
}
.highlight-header {
  color: blue;
}
.highlight-quote {
  color: #090;
}
.highlight-link {
  color: #00c;
}
/* apply the same style to codemirror */
.cm-s-ipython span.cm-keyword {
  color: #008000;
  font-weight: bold;
}
.cm-s-ipython span.cm-atom {
  color: #88F;
}
.cm-s-ipython span.cm-number {
  color: #080;
}
.cm-s-ipython span.cm-def {
  color: #00f;
}
.cm-s-ipython span.cm-variable {
  color: #000;
}
.cm-s-ipython span.cm-operator {
  color: #AA22FF;
  font-weight: bold;
}
.cm-s-ipython span.cm-variable-2 {
  color: #1a1a1a;
}
.cm-s-ipython span.cm-variable-3 {
  color: #333333;
}
.cm-s-ipython span.cm-comment {
  color: #408080;
  font-style: italic;
}
.cm-s-ipython span.cm-string {
  color: #BA2121;
}
.cm-s-ipython span.cm-string-2 {
  color: #f50;
}
.cm-s-ipython span.cm-meta {
  color: #AA22FF;
}
.cm-s-ipython span.cm-qualifier {
  color: #555;
}
.cm-s-ipython span.cm-builtin {
  color: #008000;
}
.cm-s-ipython span.cm-bracket {
  color: #997;
}
.cm-s-ipython span.cm-tag {
  color: #170;
}
.cm-s-ipython span.cm-attribute {
  color: #00c;
}
.cm-s-ipython span.cm-header {
  color: blue;
}
.cm-s-ipython span.cm-quote {
  color: #090;
}
.cm-s-ipython span.cm-link {
  color: #00c;
}
.cm-s-ipython span.cm-error {
  color: #f00;
}
.cm-s-ipython span.cm-tab {
  background: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADAAAAAMCAYAAAAkuj5RAAAAAXNSR0IArs4c6QAAAGFJREFUSMft1LsRQFAQheHPowAKoACx3IgEKtaEHujDjORSgWTH/ZOdnZOcM/sgk/kFFWY0qV8foQwS4MKBCS3qR6ixBJvElOobYAtivseIE120FaowJPN75GMu8j/LfMwNjh4HUpwg4LUAAAAASUVORK5CYII=);
  background-position: right;
  background-repeat: no-repeat;
}
div.output_wrapper {
  /* this position must be relative to enable descendents to be absolute within it */
  position: relative;
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: vertical;
  -moz-box-align: stretch;
  display: box;
  box-orient: vertical;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: column;
  align-items: stretch;
  z-index: 1;
}
/* class for the output area when it should be height-limited */
div.output_scroll {
  /* ideally, this would be max-height, but FF barfs all over that */
  height: 24em;
  /* FF needs this *and the wrapper* to specify full width, or it will shrinkwrap */
  width: 100%;
  overflow: auto;
  border-radius: 2px;
  -webkit-box-shadow: inset 0 2px 8px rgba(0, 0, 0, 0.8);
  box-shadow: inset 0 2px 8px rgba(0, 0, 0, 0.8);
  display: block;
}
/* output div while it is collapsed */
div.output_collapsed {
  margin: 0px;
  padding: 0px;
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: vertical;
  -moz-box-align: stretch;
  display: box;
  box-orient: vertical;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: column;
  align-items: stretch;
}
div.out_prompt_overlay {
  height: 100%;
  padding: 0px 0.4em;
  position: absolute;
  border-radius: 2px;
}
div.out_prompt_overlay:hover {
  /* use inner shadow to get border that is computed the same on WebKit/FF */
  -webkit-box-shadow: inset 0 0 1px #000;
  box-shadow: inset 0 0 1px #000;
  background: rgba(240, 240, 240, 0.5);
}
div.output_prompt {
  color: #D84315;
}
/* This class is the outer container of all output sections. */
div.output_area {
  padding: 0px;
  page-break-inside: avoid;
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: horizontal;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: horizontal;
  -moz-box-align: stretch;
  display: box;
  box-orient: horizontal;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: row;
  align-items: stretch;
}
div.output_area .MathJax_Display {
  text-align: left !important;
}
div.output_area 
div.output_area 
div.output_area img,
div.output_area svg {
  max-width: 100%;
  height: auto;
}
div.output_area img.unconfined,
div.output_area svg.unconfined {
  max-width: none;
}
/* This is needed to protect the pre formating from global settings such
   as that of bootstrap */
.output {
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: vertical;
  -moz-box-align: stretch;
  display: box;
  box-orient: vertical;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: column;
  align-items: stretch;
}
@media (max-width: 540px) {
  div.output_area {
    /* Old browsers */
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-box-align: stretch;
    display: -moz-box;
    -moz-box-orient: vertical;
    -moz-box-align: stretch;
    display: box;
    box-orient: vertical;
    box-align: stretch;
    /* Modern browsers */
    display: flex;
    flex-direction: column;
    align-items: stretch;
  }
}
div.output_area pre {
  margin: 0;
  padding: 0;
  border: 0;
  vertical-align: baseline;
  color: black;
  background-color: transparent;
  border-radius: 0;
}
/* This class is for the output subarea inside the output_area and after
   the prompt div. */
div.output_subarea {
  overflow-x: auto;
  padding: 0.4em;
  /* Old browsers */
  -webkit-box-flex: 1;
  -moz-box-flex: 1;
  box-flex: 1;
  /* Modern browsers */
  flex: 1;
  max-width: calc(100% - 14ex);
}
div.output_scroll div.output_subarea {
  overflow-x: visible;
}
/* The rest of the output_* classes are for special styling of the different
   output types */
/* all text output has this class: */
div.output_text {
  text-align: left;
  color: #000;
  /* This has to match that of the the CodeMirror class line-height below */
  line-height: 1.21429em;
}
/* stdout/stderr are 'text' as well as 'stream', but execute_result/error are *not* streams */
div.output_stderr {
  background: #fdd;
  /* very light red background for stderr */
}
div.output_latex {
  text-align: left;
}
/* Empty output_javascript divs should have no height */
div.output_javascript:empty {
  padding: 0;
}
.js-error {
  color: darkred;
}
/* raw_input styles */
div.raw_input_container {
  line-height: 1.21429em;
  padding-top: 5px;
}
pre.raw_input_prompt {
  /* nothing needed here. */
}
input.raw_input {
  font-family: monospace;
  font-size: inherit;
  color: inherit;
  width: auto;
  /* make sure input baseline aligns with prompt */
  vertical-align: baseline;
  /* padding + margin = 0.5em between prompt and cursor */
  padding: 0em 0.25em;
  margin: 0em 0.25em;
}
input.raw_input:focus {
  box-shadow: none;
}
p.p-space {
  margin-bottom: 10px;
}
div.output_unrecognized {
  padding: 5px;
  font-weight: bold;
  color: red;
}
div.output_unrecognized a {
  color: inherit;
  text-decoration: none;
}
div.output_unrecognized a:hover {
  color: inherit;
  text-decoration: none;
}
.rendered_html {
  color: #000;
  /* any extras will just be numbers: */
}



.rendered_html :link {
  text-decoration: underline;
}
.rendered_html :visited {
  text-decoration: underline;
}






.rendered_html h1:first-child {
  margin-top: 0.538em;
}
.rendered_html h2:first-child {
  margin-top: 0.636em;
}
.rendered_html h3:first-child {
  margin-top: 0.777em;
}
.rendered_html h4:first-child {
  margin-top: 1em;
}
.rendered_html h5:first-child {
  margin-top: 1em;
}
.rendered_html h6:first-child {
  margin-top: 1em;
}








.rendered_html * + ul {
  margin-top: 1em;
}
.rendered_html * + ol {
  margin-top: 1em;
}


.rendered_html pre,



.rendered_html tr,
.rendered_html th,

.rendered_html td,


.rendered_html * + table {
  margin-top: 1em;
}

.rendered_html * + p {
  margin-top: 1em;
}

.rendered_html * + img {
  margin-top: 1em;
}
.rendered_html img,

.rendered_html img.unconfined,

div.text_cell {
  /* Old browsers */
  display: -webkit-box;
  -webkit-box-orient: horizontal;
  -webkit-box-align: stretch;
  display: -moz-box;
  -moz-box-orient: horizontal;
  -moz-box-align: stretch;
  display: box;
  box-orient: horizontal;
  box-align: stretch;
  /* Modern browsers */
  display: flex;
  flex-direction: row;
  align-items: stretch;
}
@media (max-width: 540px) {
  div.text_cell > div.prompt {
    display: none;
  }
}
div.text_cell_render {
  /*font-family: "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;*/
  outline: none;
  resize: none;
  width: inherit;
  border-style: none;
  padding: 0.5em 0.5em 0.5em 0.4em;
  color: #000;
  box-sizing: border-box;
  -moz-box-sizing: border-box;
  -webkit-box-sizing: border-box;
}
a.anchor-link:link {
  text-decoration: none;
  padding: 0px 20px;
  visibility: hidden;
}
h1:hover .anchor-link,
h2:hover .anchor-link,
h3:hover .anchor-link,
h4:hover .anchor-link,
h5:hover .anchor-link,
h6:hover .anchor-link {
  visibility: visible;
}
.text_cell.rendered .input_area {
  display: none;
}
.text_cell.rendered 
.text_cell.unrendered .text_cell_render {
  display: none;
}
.cm-header-1,
.cm-header-2,
.cm-header-3,
.cm-header-4,
.cm-header-5,
.cm-header-6 {
  font-weight: bold;
  font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
}
.cm-header-1 {
  font-size: 185.7%;
}
.cm-header-2 {
  font-size: 157.1%;
}
.cm-header-3 {
  font-size: 128.6%;
}
.cm-header-4 {
  font-size: 110%;
}
.cm-header-5 {
  font-size: 100%;
  font-style: italic;
}
.cm-header-6 {
  font-size: 100%;
  font-style: italic;
}
</style>
<style type="text/css">.highlight .hll { background-color: #ffffcc }
.highlight  { background: #f8f8f8; }
.highlight .c { color: #408080; font-style: italic } /* Comment */
.highlight .err { border: 1px solid #FF0000 } /* Error */
.highlight .k { color: #008000; font-weight: bold } /* Keyword */
.highlight .o { color: #666666 } /* Operator */
.highlight .ch { color: #408080; font-style: italic } /* Comment.Hashbang */
.highlight .cm { color: #408080; font-style: italic } /* Comment.Multiline */
.highlight .cp { color: #BC7A00 } /* Comment.Preproc */
.highlight .cpf { color: #408080; font-style: italic } /* Comment.PreprocFile */
.highlight .c1 { color: #408080; font-style: italic } /* Comment.Single */
.highlight .cs { color: #408080; font-style: italic } /* Comment.Special */
.highlight .gd { color: #A00000 } /* Generic.Deleted */
.highlight .ge { font-style: italic } /* Generic.Emph */
.highlight .gr { color: #FF0000 } /* Generic.Error */
.highlight .gh { color: #000080; font-weight: bold } /* Generic.Heading */
.highlight .gi { color: #00A000 } /* Generic.Inserted */
.highlight .go { color: #888888 } /* Generic.Output */
.highlight .gp { color: #000080; font-weight: bold } /* Generic.Prompt */
.highlight .gs { font-weight: bold } /* Generic.Strong */
.highlight .gu { color: #800080; font-weight: bold } /* Generic.Subheading */
.highlight .gt { color: #0044DD } /* Generic.Traceback */
.highlight .kc { color: #008000; font-weight: bold } /* Keyword.Constant */
.highlight .kd { color: #008000; font-weight: bold } /* Keyword.Declaration */
.highlight .kn { color: #008000; font-weight: bold } /* Keyword.Namespace */
.highlight .kp { color: #008000 } /* Keyword.Pseudo */
.highlight .kr { color: #008000; font-weight: bold } /* Keyword.Reserved */
.highlight .kt { color: #B00040 } /* Keyword.Type */
.highlight .m { color: #666666 } /* Literal.Number */
.highlight .s { color: #BA2121 } /* Literal.String */
.highlight .na { color: #7D9029 } /* Name.Attribute */
.highlight .nb { color: #008000 } /* Name.Builtin */
.highlight .nc { color: #0000FF; font-weight: bold } /* Name.Class */
.highlight .no { color: #880000 } /* Name.Constant */
.highlight .nd { color: #AA22FF } /* Name.Decorator */
.highlight .ni { color: #999999; font-weight: bold } /* Name.Entity */
.highlight .ne { color: #D2413A; font-weight: bold } /* Name.Exception */
.highlight .nf { color: #0000FF } /* Name.Function */
.highlight .nl { color: #A0A000 } /* Name.Label */
.highlight .nn { color: #0000FF; font-weight: bold } /* Name.Namespace */
.highlight .nt { color: #008000; font-weight: bold } /* Name.Tag */
.highlight .nv { color: #19177C } /* Name.Variable */
.highlight .ow { color: #AA22FF; font-weight: bold } /* Operator.Word */
.highlight .w { color: #bbbbbb } /* Text.Whitespace */
.highlight .mb { color: #666666 } /* Literal.Number.Bin */
.highlight .mf { color: #666666 } /* Literal.Number.Float */
.highlight .mh { color: #666666 } /* Literal.Number.Hex */
.highlight .mi { color: #666666 } /* Literal.Number.Integer */
.highlight .mo { color: #666666 } /* Literal.Number.Oct */
.highlight .sa { color: #BA2121 } /* Literal.String.Affix */
.highlight .sb { color: #BA2121 } /* Literal.String.Backtick */
.highlight .sc { color: #BA2121 } /* Literal.String.Char */
.highlight .dl { color: #BA2121 } /* Literal.String.Delimiter */
.highlight .sd { color: #BA2121; font-style: italic } /* Literal.String.Doc */
.highlight .s2 { color: #BA2121 } /* Literal.String.Double */
.highlight .se { color: #BB6622; font-weight: bold } /* Literal.String.Escape */
.highlight .sh { color: #BA2121 } /* Literal.String.Heredoc */
.highlight .si { color: #BB6688; font-weight: bold } /* Literal.String.Interpol */
.highlight .sx { color: #008000 } /* Literal.String.Other */
.highlight .sr { color: #BB6688 } /* Literal.String.Regex */
.highlight .s1 { color: #BA2121 } /* Literal.String.Single */
.highlight .ss { color: #19177C } /* Literal.String.Symbol */
.highlight .bp { color: #008000 } /* Name.Builtin.Pseudo */
.highlight .fm { color: #0000FF } /* Name.Function.Magic */
.highlight .vc { color: #19177C } /* Name.Variable.Class */
.highlight .vg { color: #19177C } /* Name.Variable.Global */
.highlight .vi { color: #19177C } /* Name.Variable.Instance */
.highlight .vm { color: #19177C } /* Name.Variable.Magic */
.highlight .il { color: #666666 } /* Literal.Number.Integer.Long */</style>
<style type="text/css">
/* Temporary definitions which will become obsolete with Notebook release 5.0 */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-bold { font-weight: bold; }
</style>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Today I was wondering, given a set of reads and a genome sequence, how I could filter out all of the reads that align to the genome perfectly.
I'll explore this idea a bit here.
I'm going to use <a href="https://github.com/dib-lab/nullgraph">nullgraph</a> to simulate a random genome, <a href="https://github.com/lh3/bwa">bwa</a> to compute the alignments, and <a href="http://www.htslib.org/">samtools</a> for post-processing the alignments.</p>
<h2 id="Simulated-data">Simulated data<a class="anchor-link" href="#Simulated-data">&#182;</a></h2><p>I'll start by simulating a random 2.5 megabase genome and "sequencing" to a depth of 20x coverage.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-bash"><pre><span></span>make-random-genome.py -h
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>usage: make-random-genome.py [-h] [-l LENGTH] [-s SEED] [--name NAME]

optional arguments:
  -h, --help            show this help message and exit
  -l LENGTH, --length LENGTH
                        Simulated genome length
  -s SEED, --seed SEED  Random number seed
  --name NAME           sequence name
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[2]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-bash"><pre><span></span><span class="c1"># Generating the bogus chromosome in 10 smaller pieces is quicker than in one big piece.</span>
<span class="nb">echo</span> <span class="s1">&#39;&gt;bogusgenome&#39;</span> &gt; bogus-genome.fa
<span class="nb">time</span> <span class="k">for</span> i in <span class="o">{</span><span class="m">1</span>..10<span class="o">}</span>
<span class="k">do</span>
    make-random-genome.py --length <span class="m">250000</span> --seed <span class="k">$((</span><span class="m">42</span> <span class="o">*</span> <span class="nv">$i</span><span class="k">))</span> <span class="p">|</span> grep -v <span class="s1">&#39;^&gt;&#39;</span> &gt;&gt; bogus-genome.fa
<span class="k">done</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>Using random seed: 42
Using random seed: 84
Using random seed: 126
Using random seed: 168
Using random seed: 210
Using random seed: 252
Using random seed: 294
Using random seed: 336
Using random seed: 378
Using random seed: 420

real	0m3.067s
user	0m2.876s
sys	0m0.143s
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[3]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-bash"><pre><span></span>make-reads.py -h
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>usage: make-reads.py [-h] [-e ERROR_RATE] [-r READ_LENGTH] [-C COVERAGE]
                     [-S SEED] [--even] [--mutation-details MUTATION_DETAILS]
                     genome

positional arguments:
  genome

optional arguments:
  -h, --help            show this help message and exit
  -e ERROR_RATE, --error-rate ERROR_RATE
  -r READ_LENGTH, --read-length READ_LENGTH
                        Length of reads to generate
  -C COVERAGE, --coverage COVERAGE
                        Targeted coverage level
  -S SEED, --seed SEED  Random seed
  --even
  --mutation-details MUTATION_DETAILS
                        Write detailed log of mutations here
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[4]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-bash"><pre><span></span><span class="nb">time</span> make-reads.py --error-rate <span class="m">0</span>.005 --read-length <span class="m">125</span> --coverage <span class="m">20</span> --seed <span class="m">54321</span> bogus-genome.fa &gt; bogus-reads.fa
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>genome size: 2500000
coverage: 20.0
readlen: 125
error rate: 0.005
Read in template genome bogusgenome of length 2500000 from bogus-genome.fa
Generating 400000 reads of length 125 for a target coverage of 20.0 with a target error rate of 0.005
186317 of 400000 reads mutated; 251371 total mutations

real	1m39.693s
user	1m39.133s
sys	0m0.283s
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Ok, so we have 400k reads, 125 bp each and 186,317 of these reads have one or more "mutations" (representing single-nucleotide sequencing errors).</p>
<h2 id="Take-Un">Take Un<a class="anchor-link" href="#Take-Un">&#182;</a></h2><p>My first thought was to take the "minimum seed length" parameter to the extreme.
Many alignment algorithms, <strong>bwa mem</strong> included, use a "seed-and-extend" strategy.
The <em>seed</em> is a subsequence of the read that must match the reference perfectly for the rest of the mismatch-tolerant alignment to be computed.
If my reads are 125 bp in length and I set the minimum seed length to 125, only alignments for reads that match perfectly should be reported.
The remainder of the reads should be reported as unaligned.</p>
<p>Let's see!</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[5]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-bash"><pre><span></span><span class="c1"># First, gotta index the genome so bwa can do its thang.</span>
bwa index bogus-genome.fa
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>[bwa_index] Pack FASTA... 0.03 sec
[bwa_index] Construct BWT for the packed sequence...
[bwa_index] 0.83 seconds elapse.
[bwa_index] Update BWT... 0.03 sec
[bwa_index] Pack forward-only FASTA... 0.02 sec
[bwa_index] Construct SA from BWT and Occ... 0.20 sec
[main] Version: 0.7.15-r1140
[main] CMD: bwa index bogus-genome.fa
[main] Real time: 1.122 sec; CPU: 1.113 sec
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>I'll pipe the output of <code>bwa</code> directly to <code>samtools</code>.
The <code>-f 4</code> flag will pull out all reads that are marked as unaligned, and the <code>-c</code> flag will count the records instead of printing them.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[6]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-bash"><pre><span></span>bwa mem -k <span class="m">125</span> bogus-genome.fa bogus-reads.fa <span class="p">|</span> samtools view -f <span class="m">4</span> -c
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 80000 sequences (10000000 bp)...
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 2.159 CPU sec, 2.105 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 1.929 CPU sec, 1.812 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 1.971 CPU sec, 1.857 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 2.274 CPU sec, 2.168 real sec
[M::mem_process_seqs] Processed 80000 reads in 1.830 CPU sec, 1.759 real sec
[main] Version: 0.7.15-r1140
[main] CMD: bwa mem -k 125 bogus-genome.fa bogus-reads.fa
[main] Real time: 9.840 sec; CPU: 10.281 sec
186317
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Great, it worked!
We have 186,317 reads marked as unaligned, which matches the number of reads containing errors from our simulation.</p>
<h2 id="Take-Deux">Take Deux<a class="anchor-link" href="#Take-Deux">&#182;</a></h2><p>That was pretty quick and easy, but unfortunately this solution has limited utility.
If any kind of trimming, error correction, or other quality control is performed on the reads prior to alignment, then it's almost certain that the reads will not be of uniform length.
If I set the minimum seed length to high, some reads will not make the cut simply because they aren't long enough, but if I set it too low, reads with errors will be aligned and I'll have to resort to some other way of distinguishing perfect matches anyway.</p>
<p>My next thought was to adjust the alignment scoring scheme and penalize mismatches, gap opens, and gap extensions so much that a read with any mismatch would not be aligned.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[7]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-bash"><pre><span></span>bwa mem -B <span class="m">100</span> -O <span class="o">[</span><span class="m">100</span>,100<span class="o">]</span> -E <span class="o">[</span><span class="m">100</span>,100<span class="o">]</span> bogus-genome.fa bogus-reads.fa <span class="p">|</span> samtools view -f <span class="m">4</span> -c
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 80000 sequences (10000000 bp)...
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 3.261 CPU sec, 3.207 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 7.165 CPU sec, 7.206 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 5.006 CPU sec, 5.368 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 4.185 CPU sec, 4.146 real sec
[M::mem_process_seqs] Processed 80000 reads in 6.265 CPU sec, 6.539 real sec
[main] Version: 0.7.15-r1140
[main] CMD: bwa mem -B 100 -O [100,100] -E [100,100] bogus-genome.fa bogus-reads.fa
[main] Real time: 26.631 sec; CPU: 26.008 sec
7
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[8]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-bash"><pre><span></span>bwa mem -B <span class="m">1000</span> -O <span class="o">[</span><span class="m">1000</span>,1000<span class="o">]</span> -E <span class="o">[</span><span class="m">1000</span>,1000<span class="o">]</span> bogus-genome.fa bogus-reads.fa <span class="p">|</span> samtools view -f <span class="m">4</span> -c
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 80000 sequences (10000000 bp)...
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 4.372 CPU sec, 4.334 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 4.562 CPU sec, 4.719 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 6.073 CPU sec, 6.102 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 3.534 CPU sec, 3.410 real sec
[M::mem_process_seqs] Processed 80000 reads in 3.393 CPU sec, 3.320 real sec
[main] Version: 0.7.15-r1140
[main] CMD: bwa mem -B 1000 -O [1000,1000] -E [1000,1000] bogus-genome.fa bogus-reads.fa
[main] Real time: 22.047 sec; CPU: 22.072 sec
0
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Not sure what's going on here, but clearly it didn't work as intended.</p>
<h2 id="Take-Trois">Take Trois<a class="anchor-link" href="#Take-Trois">&#182;</a></h2><p>My next thought was to process the alignments "normally" (with default parameters) and examine each alignment's CIGAR string to distinguish perfect matches from imperfect alignments.
Ah, if only life were that easy.
The CIGAR format uses different symbols to annotated insertions, deletions, clipped alignments, and so on.
But someone somwhere a long time ago decided that it would be a good idea to use the same symbol to annotate matches and mismatches in an alignment.
So if an alignment has a CIGAR string of <code>108M</code> it's impossible to know if that's a perfect match of 108 nucleotides, or 90 matches and then a single mismatch and then 17 more matches, or something else.
One would ether have to compare the read sequence to the sequence of the genome to which it was aligned, or compare the length of the read to the number of matches in the <a href="http://bio-bwa.sourceforge.net/bwa.shtml#4">MD tag</a> or something like that.
This is getting into the realm of BAM processing libraries and scripts, which I'm trying to avoid.
If at all possible I'd love to do this as a simple shell one-liner.</p>
<p>My friend <a href="https://twitter.com/srihari176">Srihari</a> suggested I look at the optional tags in more detail, particularly the <a href="http://bio-bwa.sourceforge.net/bwa.shtml#4">edit distance (NM) tag</a>.
I felt silly for not thinking of this earlier myself (I'll chalk it up to my limited practical experience with SAM/BAM data).
Any perfect alignment is going to have an edit distance of 0, while any imperfect alignment is going to have a positive edit distance, so that should be a simple way to distinguish and ignore perfect alignments.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[9]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-bash"><pre><span></span><span class="c1"># The grep command discards read alignments with an edit distance of 0.</span>
bwa mem bogus-genome.fa bogus-reads.fa <span class="p">|</span> grep -v <span class="s1">&#39;NM:i:0&#39;</span> <span class="p">|</span> wc -l
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 80000 sequences (10000000 bp)...
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 3.601 CPU sec, 3.550 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 3.519 CPU sec, 3.405 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 5.629 CPU sec, 5.714 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 5.022 CPU sec, 5.151 real sec
[M::mem_process_seqs] Processed 80000 reads in 3.737 CPU sec, 3.669 real sec
[main] Version: 0.7.15-r1140
[main] CMD: bwa mem bogus-genome.fa bogus-reads.fa
[main] Real time: 21.635 sec; CPU: 21.627 sec
  186194
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Huh, that's close, but not exactly our target.
We should be seeing 186,317 reads.
What's going on?
Let's do a bit of detective work.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[10]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-bash"><pre><span></span><span class="c1"># Grab read IDs from the &quot;seed length&quot; method that gave us the correct answer earlier.</span>
bwa mem -k <span class="m">125</span> bogus-genome.fa bogus-reads.fa <span class="p">|</span> samtools view -f <span class="m">4</span> &gt; test-one.sam
cut -f <span class="m">1</span> test-one.sam <span class="p">|</span> sort <span class="p">|</span> uniq &gt; test-one.txt

<span class="c1"># Grab read IDs from the &quot;edit distance&quot; method that isn&#39;t quite right.</span>
bwa mem bogus-genome.fa bogus-reads.fa <span class="p">|</span> samtools view <span class="p">|</span> grep -v <span class="s1">&#39;NM:i:0&#39;</span> &gt; test-two.sam
cut -f <span class="m">1</span> test-two.sam <span class="p">|</span> sort <span class="p">|</span> uniq &gt; test-two.txt

<span class="c1"># Randomly grab a few reads that are appear in the first set but not the second.</span>
<span class="c1"># I&#39;m using `gshuf` on Mac OS X. Replace it with `shuf` if you&#39;re using a Linux OS.</span>
comm -23 test-one.txt test-two.txt <span class="p">|</span> gshuf <span class="p">|</span> head -n <span class="m">5</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 80000 sequences (10000000 bp)...
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 1.816 CPU sec, 1.763 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 2.448 CPU sec, 2.477 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 2.712 CPU sec, 2.592 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 1.967 CPU sec, 1.850 real sec
[M::mem_process_seqs] Processed 80000 reads in 2.266 CPU sec, 2.218 real sec
[main] Version: 0.7.15-r1140
[main] CMD: bwa mem -k 125 bogus-genome.fa bogus-reads.fa
[main] Real time: 11.074 sec; CPU: 11.341 sec
[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 80000 sequences (10000000 bp)...
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 3.708 CPU sec, 3.658 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 4.044 CPU sec, 3.941 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 4.456 CPU sec, 4.382 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 4.368 CPU sec, 4.279 real sec
[M::mem_process_seqs] Processed 80000 reads in 4.595 CPU sec, 4.713 real sec
[main] Version: 0.7.15-r1140
[main] CMD: bwa mem bogus-genome.fa bogus-reads.fa
[main] Real time: 21.216 sec; CPU: 21.325 sec
read108697r
read74495f
read264906r
read294922f
read306338r
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Ok let's take a closer look at a couple of these reads.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[11]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-bash"><pre><span></span>grep -A <span class="m">1</span> -e read43849r -e read5193r bogus-reads.fa
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>&gt;read5193r start=1764079,mutations=2
GAAGCGATAGATCCACGCGATACTCAGTGTATGTAGCCAAATCAAAAAACTACACGATGCCGAACAGAGTGTTTTATTTAACCTCCTGTTTTTGTACAACGTTTTGAGAACATATATCATgGaTT
--
&gt;read43849r start=2013628,mutations=2
aaTTCGCGATTCTAAGGCGCGGGGTCACCATTTAATAACCTCGCACTCGGTACTGAAGTAGCGGCCGTTCGTTTGTGGGCACTAGGTCATTCAGTCCCGTGGCGCGTTTTATAGACGAAACTTTA
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Interesting.
Both have a couple of sequencing errors near the end of the read (shown as lowercase symbols).
Now let's take a look at the alignments without any filtering.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[12]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-bash"><pre><span></span><span class="c1"># Hide the progress reports by sending them to /dev/null</span>
bwa mem bogus-genome.fa bogus-reads.fa <span class="m">2</span>&gt; /dev/null <span class="p">|</span> grep -e read43849r -e read5193r
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>read5193r	16	bogusgenome	1764085	60	5S120M	*	0	0	AATCCATGATATATGTTCTCAAAACGTTGTACAAAAACAGGAGGTTAAATAAAACACTCTGTTCGGCATCGTGTAGTTTTTTGATTTGGCTACATACACTGAGTATCGCGTGGATCTATCGCTTC	*	NM:i:0	MD:Z:120	AS:i:120	XS:i:0
read43849r	16	bogusgenome	2013629	60	123M2S	*	0	0	TAAAGTTTCGTCTATAAAACGCGCCACGGGACTGAATGACCTAGTGCCCACAAACGAACGGCCGCTACTTCAGTACCGAGTGCGAGGTTATTAAATGGTGACCCCGCGCCTTAGAATCGCGAATT	*	NM:i:0	MD:Z:123	AS:i:123	XS:i:0
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Ahh, ok.
As the CIGAR strings show, both of these alignments are soft-clipped at the ends.
That means the offending nucleotides are not considered part of the alignment, e.g. when computing the edit distance.
Ignoring the soft-clipped portion, the remainder of the read is a perfect match and is thus being (erroneously) discarded.</p>
<h2 id="Take-Quatre">Take Quatre<a class="anchor-link" href="#Take-Quatre">&#182;</a></h2><p>Ok, so it looks like we need to check for soft clipping before we apply the edit distance filter.
It doesn't look like soft clipping can be disabled with <strong>bwa mem</strong> (I would LOVE to be corrected!), so it'll have to be handled by post-processing.</p>
<p>Our algorithm boils down to this: keep the record if the CIGAR string does not match the regular expression <code>\d+M</code> (check for soft clipping) or if the edit distance is 0.
I can think of two ways of doing this.
The first approach is lazier, but is also more concise and less cryptic to those who have never slung Perl.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[13]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-bash"><pre><span></span>bwa mem bogus-genome.fa bogus-reads.fa <span class="p">|</span> samtools view <span class="p">|</span> perl -ne <span class="s1">&#39;print if !/\t\d+M\t/ || !/NM:i:0/&#39;</span> <span class="p">|</span> wc -l
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 80000 sequences (10000000 bp)...
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 5.212 CPU sec, 5.262 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 6.921 CPU sec, 6.943 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 6.100 CPU sec, 6.010 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 3.542 CPU sec, 3.420 real sec
[M::mem_process_seqs] Processed 80000 reads in 3.608 CPU sec, 3.531 real sec
[main] Version: 0.7.15-r1140
[main] CMD: bwa mem bogus-genome.fa bogus-reads.fa
[main] Real time: 25.384 sec; CPU: 25.531 sec
  186317
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The second approach is more explicit, avoiding any unintended off-target matches for the <code>\d+M</code> regex.
This comes at the cost of introducing more complexity and verbosity in the Perl code, and honestly I don't know if an off-target match for that regex is even possible for well-formatted SAM alignments.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[14]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-bash"><pre><span></span>bwa mem bogus-genome.fa bogus-reads.fa <span class="p">|</span> samtools view <span class="p">|</span> perl -ne <span class="s1">&#39;@v = split; print if $v[5] =~ /S/ || !/NM:i:0/&#39;</span> <span class="p">|</span> wc -l
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 80000 sequences (10000000 bp)...
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 3.712 CPU sec, 3.667 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 5.299 CPU sec, 5.271 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 3.721 CPU sec, 3.590 real sec
[M::process] read 80000 sequences (10000000 bp)...
[M::mem_process_seqs] Processed 80000 reads in 4.195 CPU sec, 4.070 real sec
[M::mem_process_seqs] Processed 80000 reads in 4.451 CPU sec, 4.383 real sec
[main] Version: 0.7.15-r1140
[main] CMD: bwa mem bogus-genome.fa bogus-reads.fa
[main] Real time: 21.338 sec; CPU: 21.505 sec
  186317
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>In any case, both approaches see to work correctly and give us just the reads that do not align perfectly to the reference.</p>
<h2 id="Fin">Fin<a class="anchor-link" href="#Fin">&#182;</a></h2><p>One thing I haven't considered much here is performance.
This whole exercise would have been pretty silly if a short read aligner exists that make it easy to require only exact alignments to be reported.
But I'm not familiar with any, and this use case has such limited utility with the data we work with on a daily basis that I doubt there is one in wide use.
Again, I'd be happy to be proven wrong!</p>

</div>
</div>
</div>
 


<script type="text/javascript">if (!document.getElementById('mathjaxscript_pelican_#%@#$@#')) {
    var mathjaxscript = document.createElement('script');
    mathjaxscript.id = 'mathjaxscript_pelican_#%@#$@#';
    mathjaxscript.type = 'text/javascript';
    mathjaxscript.src = '//cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML';
    mathjaxscript[(window.opera ? "innerHTML" : "text")] =
        "MathJax.Hub.Config({" +
        "    config: ['MMLorHTML.js']," +
        "    TeX: { extensions: ['AMSmath.js','AMSsymbols.js','noErrors.js','noUndefined.js'], equationNumbers: { autoNumber: 'AMS' } }," +
        "    jax: ['input/TeX','input/MathML','output/HTML-CSS']," +
        "    extensions: ['tex2jax.js','mml2jax.js','MathMenu.js','MathZoom.js']," +
        "    displayAlign: 'center'," +
        "    displayIndent: '0em'," +
        "    showMathMenu: true," +
        "    tex2jax: { " +
        "        inlineMath: [ ['$','$'] ], " +
        "        displayMath: [ ['$$','$$'] ]," +
        "        processEscapes: true," +
        "        preview: 'TeX'," +
        "    }, " +
        "    'HTML-CSS': { " +
        "        styles: { '.MathJax_Display, .MathJax .mo, .MathJax .mi, .MathJax .mn': {color: 'black ! important'} }" +
        "    } " +
        "}); ";
    (document.body || document.getElementsByTagName('head')[0]).appendChild(mathjaxscript);
}
</script>
</div>
		<div>
			<h2>Comments</h2>
<div id="disqus_thread"></div>		<div>
	</div>
		<footer>
		  <p>
			<a href="http://creativecommons.org/licenses/by/4.0/"><img src="https://standage.github.io/images/cc-by-88x31.png" alt="CC BY 4.0" /></a><br />
		    &copy; 2015-2016 Daniel S. Standage.<br />
			Unless otherwise noted, all content on this site is posted under a <a href="http://creativecommons.org/licenses/by/4.0/">CC BY license</a>.</p>
		  <p></p>
		  <p>This site is powered by <a href="http://getpelican.com/">Pelican</a>, its theme based on <a href="http://twitter.github.com/bootstrap/">Twitter Bootstrap</a>.</p>
		</footer>
	  </div>

	</div>
</body>
</html>