<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8">
	<title>Daniel S. Standage</title>
	<meta name="description" content="">
	<meta name="author" content="Daniel S. Standage">

	<!-- HTML5 shim, for IE6-8 support of HTML elements -->
	<!--[if lt IE 9]>
		<script src="https://standage.github.io/theme/html5.js"></script>
	<![endif]-->

	<!-- Styles -->
	<link href="https://standage.github.io/theme/bootstrap.min.css" rel="stylesheet">
	<link href="https://standage.github.io/theme/local.css" rel="stylesheet">
	<link href="https://standage.github.io/theme/pygments.css" rel="stylesheet">

	<!-- Feeds -->
	<link href="https://standage.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Daniel S. Standage Atom Feed" />


	<link href="https://standage.github.io/feed.xml" type="application/rss+xml" rel="alternate" title="Daniel S. Standage RSS Feed" />


<script type="text/javascript">
	var disqus_identifier = "drafts/novel-k-mers-as-signatures-of-germline-mutation.html";
	(function() {
		var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
		dsq.src = 'https://standage-github-io.disqus.com/embed.js';
		(document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
	})();
</script>
</head>
<body>
	<div class="topbar">
	  <div class="topbar-inner">
		<div class="container-fluid">
		  <a class="brand" href="https://standage.github.io/">Daniel S. Standage</a>
			<ul class="nav">
				<li><a href="https://scholar.google.com/citations?user=OohLwhsAAAAJ&hl=en">Publications</a></li>
				<li><a href="https://github.com/standage">Github</a></li>
				<li><a href="https://twitter.com/byuhobbes">@byuhobbes</a></li>
				<li><a href="https://standage.github.io/notebook.html">Notebook</a></li>
			</ul>
			<p class="pull-right">
				<a href="https://standage.github.io/archives.html">[archives]</a>
				<a href="https://standage.github.io/categories.html">[categories]</a>
				<a href="https://standage.github.io/tags.html">[tags]</a>
			</p>
		</div>
	  </div>
	</div>

	<div class="container-fluid">
	  <div class="sidebar">
		<div class="well">
			<img src="https://www.gravatar.com/avatar/55b2cca3b7a11df3a961d57a68705503?size=180" alt="Daniel Standage" />
			<h3>Daniel S. Standage</h3>
			<p>
				Principal Investigator,<br />
				Bioinformatics<br />
				NBACC<br />
			</p>
			<!--<h3>Affiliations</h3>
			<ul>
				<li><a href="http://ivory.idyll.org/lab/">Lab for Data Intensive Biology</a></li>
				<li><a href="http://brendelgroup.org/">Brendel Group</a></li>
				<li><a href="http://www.public.iastate.edu/~amytoth/Toth_lab/Home.html">Toth Lab</a></li>
			</ul>
			<div class="social">
			<h3>Social</h3>
			<ul>
				<li><a href="http://github.com/standage">GitHub</a></li>
				<li><a href="https://twitter.com/byuhobbes">Twitter</a></li>
				<li><a href="http://stackexchange.com/users/208980/daniel-standage?tab=accounts">StackExchange</a></li>
			</ul>
			</div>-->
		</div>
	  </div>
	  <div class="content">
	<div class='article'>
		<div class="page-header"><h1>Novel <em>k</em>-mers as signatures of germline mutation</h1></div>
		<div class="well small">Permalink: <a class="more" href="https://standage.github.io/drafts/novel-k-mers-as-signatures-of-germline-mutation.html">2018-05-15</a>
by <a class="url fn" href="https://standage.github.io/author/daniel-s-standage.html">Daniel S. Standage </a>
 in <a href="https://standage.github.io/category/blog.html">blog</a>
tags: <a href="https://standage.github.io/tag/kevlar.html">kevlar</a> <a href="https://standage.github.io/tag/variant-discovery.html">variant discovery</a> </div>
		<div><p>My recent research has been motivated by two related questions.</p>
<ol>
<li>Can we discover genetic variants without aligning reads to a reference genome?</li>
<li>Can we quickly find shared or unique variants that are unlikely to be found using any reference-based approach?</li>
</ol>
<p>Both questions speak to pragmatic concerns as well as considerations of curiosity and novelty.
On the pragmatic side, <em>poorly represented population diversity; non-model reference quality</em>.
On the curiosity side, we've suspected for a long time that variant discovery without read mapping is possible, but </p>
<p>These questions both address practical issues with variant discovery and analysis as well as </p>
<p>have an element of curiousity and novelty.
On the practical side, </p>
<p>The first question speaks to a variety of practical problems.</p>
<p>Understanding the roles genetic variation plays in human disease, traits of agricultural interest, and in the process of evolution has become an area of intense research interest.
An increasing number of fully sequ</p>
<p>Cataloging genetic variation within families and across populations has become an area of intense research interest in biomedicine, agriculture, ecology, and related fields.
D</p>
<p>Databases of genetic variants play a critical role in biomedical research, and cataloging genetic variation within families and across populations is an area of intense interest and activity.
Discovery of genetic variants is </p></div>
		<div>
			<h2>Comments</h2>
<div id="disqus_thread"></div>		<div>
	</div>
		<footer>
		  <p>
			<a href="http://creativecommons.org/licenses/by/4.0/"><img src="https://standage.github.io/images/cc-by-88x31.png" alt="CC BY 4.0" /></a><br />
		    &copy; 2015-2022 Daniel S. Standage.<br />
			Unless otherwise noted, all content on this site is posted under a <a href="http://creativecommons.org/licenses/by/4.0/">CC BY license</a>.</p>
		  <p></p>
		  <p>This site is powered by <a href="http://getpelican.com/">Pelican</a>, its theme based on <a href="http://twitter.github.com/bootstrap/">Twitter Bootstrap</a>.</p>
		</footer>
	  </div>

	</div>
</body>
</html>