<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8">
	<title>Daniel S. Standage</title>
	<meta name="description" content="">
	<meta name="author" content="Daniel S. Standage">

	<!-- HTML5 shim, for IE6-8 support of HTML elements -->
	<!--[if lt IE 9]>
		<script src="https://standage.github.io/theme/html5.js"></script>
	<![endif]-->

	<!-- Styles -->
	<link href="https://standage.github.io/theme/bootstrap.min.css" rel="stylesheet">
	<link href="https://standage.github.io/theme/local.css" rel="stylesheet">
	<link href="https://standage.github.io/theme/pygments.css" rel="stylesheet">

	<!-- Feeds -->
	<link href="https://standage.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Daniel S. Standage Atom Feed" />


	<link href="https://standage.github.io/feed.xml" type="application/rss+xml" rel="alternate" title="Daniel S. Standage RSS Feed" />


<script type="text/javascript">
	var disqus_identifier = "drafts/overview-of-hashing.html";
	(function() {
		var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
		dsq.src = 'https://standage-github-io.disqus.com/embed.js';
		(document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
	})();
</script>
</head>
<body>
	<div class="topbar">
	  <div class="topbar-inner">
		<div class="container-fluid">
		  <a class="brand" href="https://standage.github.io/">Daniel S. Standage</a>
			<ul class="nav">
				<li><a href="https://scholar.google.com/citations?user=OohLwhsAAAAJ&hl=en">Publications</a></li>
				<li><a href="https://github.com/standage">Github</a></li>
				<li><a href="https://twitter.com/byuhobbes">@byuhobbes</a></li>
				<li><a href="https://standage.github.io/notebook.html">Notebook</a></li>
			</ul>
			<p class="pull-right">
				<a href="https://standage.github.io/archives.html">[archives]</a>
				<a href="https://standage.github.io/categories.html">[categories]</a>
				<a href="https://standage.github.io/tags.html">[tags]</a>
			</p>
		</div>
	  </div>
	</div>

	<div class="container-fluid">
	  <div class="sidebar">
		<div class="well">
			<img src="https://www.gravatar.com/avatar/55b2cca3b7a11df3a961d57a68705503?size=180" alt="Daniel Standage" />
			<h3>Daniel S. Standage</h3>
			<p>
				Principal Investigator<br />
				Bioinformatics<br />
				National Bioforensic Analysis Center (NBFAC)<br />
			</p>
			<!--<h3>Affiliations</h3>
			<ul>
				<li><a href="http://ivory.idyll.org/lab/">Lab for Data Intensive Biology</a></li>
				<li><a href="http://brendelgroup.org/">Brendel Group</a></li>
				<li><a href="http://www.public.iastate.edu/~amytoth/Toth_lab/Home.html">Toth Lab</a></li>
			</ul>
			<div class="social">
			<h3>Social</h3>
			<ul>
				<li><a href="http://github.com/standage">GitHub</a></li>
				<li><a href="https://twitter.com/byuhobbes">Twitter</a></li>
				<li><a href="http://stackexchange.com/users/208980/daniel-standage?tab=accounts">StackExchange</a></li>
			</ul>
			</div>-->
		</div>
	  </div>
	  <div class="content">
	<div class='article'>
		<div class="page-header"><h1>Overview of hashing</h1></div>
		<div class="well small">Permalink: <a class="more" href="https://standage.github.io/drafts/overview-of-hashing.html">2016-09-01</a>
by <a class="url fn" href="https://standage.github.io/author/daniel-s-standage.html">Daniel S. Standage </a>
 in <a href="https://standage.github.io/category/blog.html">blog</a>
tags: <a href="https://standage.github.io/tag/hashing.html">hashing</a> <a href="https://standage.github.io/tag/k-mers.html">k-mers</a> <a href="https://standage.github.io/tag/khmer.html">khmer</a> </div>
		<div><p>Prior to my current position, I was a casual and occasional user of the <a href="https://github.com/dib-lab/khmer/">khmer library</a> for DNA sequence analysis.
In the past year and a half or so in my current position, I have become a core contributor and taken some responsibility for khmer's management and release.
During this time, we've been discussing (and to a lesser extent implementing) the integration of different <em>k</em>-mer hashing functions.
I've wanted to write blah blah.</p>
<h2>What is a hash function?</h2>
<p>The simplest useful definition of a hash function I can think of is <em>a mathematical operation that converts an arbitrary piece of data into a numerical representation</em>.
In the context of sequence analysis, the "arbitrary piece of data" is a molecular sequence (typically of a fixed length <em>k</em>; a <em>k</em>-mer) and the "numerical representation" is typically an integer value.
Several data structures common in bioinformatics use hashing to index <em>k</em>-mers for efficient storage and retrieval.</p>
<h2>Two-bit encoding</h2>
<h2>H</h2>
<p>Recently I've been digging into the internals of the <a href="https://github.com/dib-lab/khmer/">khmer package</a> and how it handles DNA sequences.
While this is all fresh on my mind, I wanted to take a moment to discuss the topic from the layman's perspective.
This post reviews the two-bit encoding scheme used by khmer, <em>k</em>-mer hashing in general, the reason khmer is limited to <em>k</em>-mers â‰¤ 32 nucleotides in length, and the implications of our plan to support larger (possibly arbitrary) values of <em>k</em>.</p>
<h2>Two-bit representation of DNA sequences</h2>
<p>The most common convention for</p>
<p>The letters A, C, G, and T have long been used as shorthand for the nucleotides from which DNA sequences are composed.</p>
<p>When stored as text, DNA sequences require 8 bits (1 byte) per nucleotide: a 32 base pair sequence requires 256 bits of storage, a 1 kilobase sequence requires 8,000 bits (1,000 bytes or 1 kilobyte), and 1 megabase requires 8,000,000 bits (1,000,000 bytes or 1 megabyte) of storage.
This has to do with how computers store information.
Each character of text requires 8 bits of information.
Consider the text.</p>
<div class="highlight"><pre><span></span><code><span class="n">|</span><span class="w">   </span><span class="n">G</span><span class="w">    </span><span class="n">|</span><span class="w">   </span><span class="n">A</span><span class="w">    </span><span class="n">|</span><span class="w">   </span><span class="n">T</span><span class="w">    </span><span class="n">|</span><span class="w">   </span><span class="n">T</span><span class="w">    </span><span class="n">|</span><span class="w">   </span><span class="n">A</span><span class="w">    </span><span class="n">|</span><span class="w">   </span><span class="n">C</span><span class="w">    </span><span class="n">|</span><span class="w">   </span><span class="n">A</span><span class="w">    </span><span class="n">|</span><span class="w"></span>
<span class="n">|01000111|01000001|01010100|01010100|01000001|01000011|01000001|</span><span class="w"></span>
</code></pre></div>

<p>However, if we hash each of these sequences to an integer value, we only require 64 bits (8 bytes) at most to store each one.
Recall that computers store information in binary format.</p>
<p>0000000000000000000000000000000000000000000000000000000000000000</p>
<p>There are a variety of ways to hash a string to an integer value.
The scheme used by khmer is based on the observation that each individual nucleotide can be unambiguously represented using only 2 bits of information.
For example, if we map the nucleotides adenine (A), thymine (T), cytosine (C), and guanine (G) to the numbers 0, 1, 2, and 3</p>
<p>nice property: perfect hash</p>
<p>When working with</p>
<div class="highlight"><pre><span></span><code>  G  |  A  |  T  |  T  |  A  |  C  |  A  |
 1 1 | 0 0 | 0 1 | 0 1 | 0 0 | 1 0 | 0 0 |
</code></pre></div>

<p>Currently, khmer only supports</p>
<p>The bulk of the <a href="https://github.com/dib-lab/khmer/">khmer package</a>'s functionality is devoted to data structures designed for querying the presence and abundance of <em>k</em>-mers (DNA sequences of a set length <em>k</em>).
khmer</p>
<p>Like many sequence analysis tools, the <a href="https://github.com/dib-lab/khmer/">khmer package</a> relies on hashing <em>k</em>-mers (DNA sequences of a set length <em>k</em>).
Defined broadly, hashing is the process by which data of arbitrary size is transformed to an alternative fixed-size representation.
Within the scope of the khmer package, <em>k</em>-mer sequences stored as text are hashed into a numerical value (an unsigned 64 bit integer, specifically) for further processing.
This transformation serves two purposes: first, the integer representation of the sequence occupies much less memory than the text-based representation; and second, the integer representation is the value used to index into various data structures for storing and querying <em>k</em>-mer presence and abundance.</p>
<p>The khmer package uses a fairly common two-bit encoding scheme
Within the scope of the khmer package, <em>k</em>-mer hashing serves primary two purposes: first, it transforms DNA sequences from a text-based representation to a numerical representation that occupies much less memory (compression); and second,</p>
<p>Generally, hashing is a transformation of arbitrary data (DNA sequence stored as plain text) to a single numerical value.</p>
<p>Hashing has two purposes: first, it transforms character/string data into a numerical representation that occupies much less memory (compression), and it</p>
<p>I've been digging</p>
<p>All of these features depend on <em>k</em>-mer hashing.</p>
<p>Broadly defined, hashing is a process by which data of arbitrary size is transformed to an alternative fixed-size representation.
In khmer, <em>k</em>-mers stored initially as text are transformed into a numerical value (an unsigned 64 bit integer) for further processing.
Storing k-mer as integers rather than as text is much more space efficient, and the integer values can be used to index directly into the hash tables used for k-mer counting.</p>
<p>Recently I've been digging into the internals of the <a href="https://github.com/dib-lab/khmer/">khmer package</a> and, in particular, its hashing scheme.
khmer uses the very common approach in bioinformatics of breaking DNA sequences into overlapping segments of length <em>k</em> called <em>k</em>-mers.
It then provides data structures for querying <em>k</em>-mer presence and abundance, as well as algorithms for data filtering and graph traversal.</p>
<p>Like many DNA sequence analysis tools, khmer uses a <em>hash function</em> to transform <em>k</em>-mers stored initially as text into a numerical value that occupies much less memory.
Currently, khmer can only support <em>k</em>-mers of length â‰¤ 32 nucleotides.
Here I discuss khmer's hashing function, the reasons for the 32 bp limit, and discuss how we might add support for larger (possibly arbitrary) value of <em>k</em>.</p></div>
		<div>
			<h2>Comments</h2>
<div id="disqus_thread"></div>		<div>
	</div>
		<footer>
		  <p>
			<a href="http://creativecommons.org/licenses/by/4.0/"><img src="https://standage.github.io/images/cc-by-88x31.png" alt="CC BY 4.0" /></a><br />
		    &copy; 2015-2022 Daniel S. Standage.<br />
			Unless otherwise noted, all content on this site is posted under a <a href="http://creativecommons.org/licenses/by/4.0/">CC BY license</a>.</p>
		  <p></p>
		  <p>This site is powered by <a href="http://getpelican.com/">Pelican</a>, its theme based on <a href="http://twitter.github.com/bootstrap/">Twitter Bootstrap</a>.</p>
		</footer>
	  </div>

	</div>
</body>
</html>