Title: Lamprey de novo transcriptome project
Date: 2016-07-18
Author: Daniel S. Standage
Category: blog
Status: draft
Tags: lamprey; rnaseq; evolution

The sea lamprey *Petromyzon marinus* is a jawless vertebrate that diverged from the gnathostome lineage (jawed vertebrates, including fish, amphibians, reptiles, and mammals) an estimated 550 million years ago.
Along with the hagfish (class Myxini), lamprey is an important system for studying basal vertebrate biology and vertebrate evolution.
The first comprehensive analysis of the lamprey genome was [published in 2013](http://dx.doi.org/doi:10.1038/ng.2568), highlighting two probable ancient whole genome duplications (predating divergence from gnathostomes) and evidence that an important limb development factor evolved within the gnathostome lineage.

Studies of lamprey at the genome level, however, are confounded by the fact that [approximately 20% of the genome is jettisoned via programmed genome rearrangements](http://dx.doi.org/10.1073/pnas.0902358106) early in development, resulting in somatic tissues lacking a complete germline genome.
As a complement to genome sequencing, [Weiming Li](https://msu.edu/~liweim/) at MSU has sequenced RNA extensively from a variety of tissues in lamprey.
Our lab has been collaborating with Weiming for several years on analysis of the lamprey transcriptome, and my work on this project will continue that collaboration.
The following questions have been investigated or are currently under investigation.

- Relationship between genome assembly and transcriptome
    - How many gene models from the genome assembly are represented in the transcriptome?
    - How many transcripts / genes from the genome assembly are mapped to the genome? How many "agree" with an annotated gene model?
    - What proportion of aligned pseudotranscripts that don't "agree" with an annotation have homology to known protein databases?
    - What is the recall of the genome annotation vs transcriptome assembly in terms of CEGs/BUSCOs?
- Novel transcripts
    - How many transcripts with open reading frames have no matches to known databases?
        - How many of these are "not short" (heuristic for "less likely to be spurious")?
        - How many of these are "long" (translated amino acid length > median; heuristic for "highly unlikely to be spurious")?
- Chimeric transcripts
    - How many can be identified?
    - How many can be corrected?
- Evolution
    - Can we identify ancestral vertebrate genes?
