<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8">
	<title>Daniel S. Standage</title>
	<meta name="description" content="">
	<meta name="author" content="Daniel S. Standage">

	<!-- HTML5 shim, for IE6-8 support of HTML elements -->
	<!--[if lt IE 9]>
		<script src="https://standage.github.io/theme/html5.js"></script>
	<![endif]-->

	<!-- Styles -->
	<link href="https://standage.github.io/theme/bootstrap.min.css" rel="stylesheet">
	<link href="https://standage.github.io/theme/local.css" rel="stylesheet">
	<link href="https://standage.github.io/theme/pygments.css" rel="stylesheet">

	<!-- Feeds -->
	<link href="https://standage.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Daniel S. Standage Atom Feed" />


	<link href="https://standage.github.io/feed.xml" type="application/rss+xml" rel="alternate" title="Daniel S. Standage RSS Feed" />


</head>
<body>
	<div class="topbar">
	  <div class="topbar-inner">
		<div class="container-fluid">
		  <a class="brand" href="https://standage.github.io/">Daniel S. Standage</a>
			<ul class="nav">
				<li><a href="https://scholar.google.com/citations?user=OohLwhsAAAAJ&hl=en">Publications</a></li>
				<li><a href="https://github.com/standage">Github</a></li>
				<li><a href="https://twitter.com/byuhobbes">@byuhobbes</a></li>
				<li><a href="https://standage.github.io/notebook.html">Notebook</a></li>
			</ul>
			<p class="pull-right">
				<a href="https://standage.github.io/archives.html">[archives]</a>
				<a href="https://standage.github.io/categories.html">[categories]</a>
				<a href="https://standage.github.io/tags.html">[tags]</a>
			</p>
		</div>
	  </div>
	</div>

	<div class="container-fluid">
	  <div class="sidebar">
		<div class="well">
			<img src="https://www.gravatar.com/avatar/55b2cca3b7a11df3a961d57a68705503?size=180" alt="Daniel Standage" />
			<h3>Daniel S. Standage</h3>
			<p>
				Assoc. Principal Investigator,<br />
				Bioinformatics<br />
				NBACC<br />
			</p>
			<!--<h3>Affiliations</h3>
			<ul>
				<li><a href="http://ivory.idyll.org/lab/">Lab for Data Intensive Biology</a></li>
				<li><a href="http://brendelgroup.org/">Brendel Group</a></li>
				<li><a href="http://www.public.iastate.edu/~amytoth/Toth_lab/Home.html">Toth Lab</a></li>
			</ul>
			<div class="social">
			<h3>Social</h3>
			<ul>
				<li><a href="http://github.com/standage">GitHub</a></li>
				<li><a href="https://twitter.com/byuhobbes">Twitter</a></li>
				<li><a href="http://stackexchange.com/users/208980/daniel-standage?tab=accounts">StackExchange</a></li>
			</ul>
			</div>-->
		</div>
	  </div>
	  <div class="content">
<div class="page-header"><h1>Notebook <small>Daniel S. Standage</small></h1></div>
<p>Psst! I've posted some details about my notebook setup <a href="https://standage.github.io/pages/notebook-setup.html">here</a>.</p>

<div class='article'>
	<h2><a href="https://standage.github.io/the-joy-and-art-of-automated-testing.html">The Joy and Art of Automated Testing</a></h2>
	<div class="well small">Permalink: <a class="more" href="https://standage.github.io/the-joy-and-art-of-automated-testing.html">2018-11-19</a>
by <a class="url fn" href="https://standage.github.io/author/daniel-s-standage.html">Daniel S. Standage </a>
 in <a href="https://standage.github.io/category/blog.html">blog</a>
tags: <a href="https://standage.github.io/tag/software.html">software</a> <a href="https://standage.github.io/tag/testing.html">testing</a> </div>
	<div class="summary"><p>A primer on software testing for scientists and researchers.</p> <a class="btn primary xsmall" href="https://standage.github.io/the-joy-and-art-of-automated-testing.html">more…</a>
	</div>
</div>	
<div class='article'>
	<h2><a href="https://standage.github.io/a-brief-review-of-hulk-and-histosketch.html">A brief review of HULK and histosketch</a></h2>
	<div class="well small">Permalink: <a class="more" href="https://standage.github.io/a-brief-review-of-hulk-and-histosketch.html">2018-10-12</a>
by <a class="url fn" href="https://standage.github.io/author/daniel-s-standage.html">Daniel S. Standage </a>
 in <a href="https://standage.github.io/category/blog.html">blog</a>
tags: <a href="https://standage.github.io/tag/sketching.html">sketching</a> <a href="https://standage.github.io/tag/metagenomics.html">metagenomics</a> <a href="https://standage.github.io/tag/journal-club.html">journal club</a> </div>
	<div class="summary"><p>About a month ago, I was intrigued to see a bit of Twitter activity around a new bioRxiv preprint.
The <a href="https://www.biorxiv.org/content/early/2018/09/04/408070">manuscript</a> describes <a href="https://github.com/will-rowe/hulk">HULK</a>, a new bioinformatics tool that implements some useful comparison metrics and operations for analyzing (meta)genomes.
HULK is based on a new algorithm called histogram sketching (HistoSketch for short), following the trend of related sketching algorithms (HyperLogLog ...</p> <a class="btn primary xsmall" href="https://standage.github.io/a-brief-review-of-hulk-and-histosketch.html">more…</a>
	</div>
</div>	
<div class='article'>
	<h2><a href="https://standage.github.io/improvements-from-applying-filters-at-k-mer-counting-time-in-kevlar.html">Improvements from applying filters at k-mer counting time in kevlar</a></h2>
	<div class="well small">Permalink: <a class="more" href="https://standage.github.io/improvements-from-applying-filters-at-k-mer-counting-time-in-kevlar.html">2018-07-16</a>
by <a class="url fn" href="https://standage.github.io/author/daniel-s-standage.html">Daniel S. Standage </a>
 in <a href="https://standage.github.io/category/blog.html">blog</a>
tags: <a href="https://standage.github.io/tag/kevlar.html">kevlar</a> </div>
	<div class="summary"><p>One of the fundamental insights of the <a href="https://kevlar.readthedocs.io">kevlar <em>de novo</em> variant caller</a> is the framing of the variant discovery problem as a search for novel k-mers.
In this case, "novel" means abundant in the focal sample and effectively absent from all control samples.
In the early stages of creating kevlar, it quickly became clear that many k-mers satisfying these simple ...</p> <a class="btn primary xsmall" href="https://standage.github.io/improvements-from-applying-filters-at-k-mer-counting-time-in-kevlar.html">more…</a>
	</div>
</div>	
<div class='article'>
	<h2><a href="https://standage.github.io/loading-paired-reads-from-position-sorted-bam-files.html">Loading paired reads from position-sorted BAM files</a></h2>
	<div class="well small">Permalink: <a class="more" href="https://standage.github.io/loading-paired-reads-from-position-sorted-bam-files.html">2018-06-12</a>
by <a class="url fn" href="https://standage.github.io/author/daniel-s-standage.html">Daniel S. Standage </a>
 in <a href="https://standage.github.io/category/blog.html">blog</a>
tags: <a href="https://standage.github.io/tag/ngs.html">ngs</a> <a href="https://standage.github.io/tag/bam.html">bam</a> </div>
	<div class="summary"><p>BAM files with sequence alignments sorted by genomic position seem to be the new currency of exchange for large-scale human genome sequencing projects.
This is convenient and practical in many ways for many people.
But in my current research I work a lot with tools that only want/need the sequence information and, for whatever reasons, support only FASTA or ...</p> <a class="btn primary xsmall" href="https://standage.github.io/loading-paired-reads-from-position-sorted-bam-files.html">more…</a>
	</div>
</div>	
<div class='article'>
	<h2><a href="https://standage.github.io/information-content-versus-data-volume-and-k-mer-counting-accuracy.html">Information content versus data volume and <em>k</em>-mer counting accuracy</a></h2>
	<div class="well small">Permalink: <a class="more" href="https://standage.github.io/information-content-versus-data-volume-and-k-mer-counting-accuracy.html">2017-10-06</a>
by <a class="url fn" href="https://standage.github.io/author/daniel-s-standage.html">Daniel S. Standage </a>
 in <a href="https://standage.github.io/category/blog.html">blog</a>
tags: <a href="https://standage.github.io/tag/kevlar.html">kevlar</a> <a href="https://standage.github.io/tag/variant-discovery.html">variant discovery</a> <a href="https://standage.github.io/tag/error-correction.html">error correction</a> </div>
	<div class="summary"><p>Keeping track of <em>k</em>-mers for simple operations has become a fundamental component of many bioinformatics techniques.
Two common operations on <em>k</em>-mers include set membership queries ("is <em>k</em>-mer <em>X</em> present in data set <em>Y</em>?") and abundance queries ("how many times does <em>k</em>-mer <em>X</em> occur in data set <em>Y</em>").
Several probabilistic data structures have been developed to support ...</p> <a class="btn primary xsmall" href="https://standage.github.io/information-content-versus-data-volume-and-k-mer-counting-accuracy.html">more…</a>
	</div>
</div>	
<div class='article'>
	<h2><a href="https://standage.github.io/how-to-distinguish-perfectly-mapped-reads-from-a-sambam-file.html">How to distinguish perfectly mapped reads from a SAM/BAM file</a></h2>
	<div class="well small">Permalink: <a class="more" href="https://standage.github.io/how-to-distinguish-perfectly-mapped-reads-from-a-sambam-file.html">2017-04-14</a>
by <a class="url fn" href="https://standage.github.io/author/daniel-s-standage.html">Daniel S. Standage </a>
 in <a href="https://standage.github.io/category/notebook.html">notebook</a>
tags: <a href="https://standage.github.io/tag/bash.html">bash</a> <a href="https://standage.github.io/tag/sam.html">sam</a> <a href="https://standage.github.io/tag/bam.html">bam</a> <a href="https://standage.github.io/tag/mapping.html">mapping</a> </div>
	<div class="summary"><p>In which I explore read alignments in SAM format and discuss the pros
and cons of various approaches to distinguishing perfect matches from
imperfect matches.</p> <a class="btn primary xsmall" href="https://standage.github.io/how-to-distinguish-perfectly-mapped-reads-from-a-sambam-file.html">more…</a>
	</div>
</div>	
<div class='article'>
	<h2><a href="https://standage.github.io/super-simple-reverse-complement-aware-dna-sequence-search-with-rcgrep.html">Super simple reverse-complement aware DNA sequence search with rcgrep</a></h2>
	<div class="well small">Permalink: <a class="more" href="https://standage.github.io/super-simple-reverse-complement-aware-dna-sequence-search-with-rcgrep.html">2017-03-02</a>
by <a class="url fn" href="https://standage.github.io/author/daniel-s-standage.html">Daniel S. Standage </a>
 in <a href="https://standage.github.io/category/blog.html">blog</a>
tags: <a href="https://standage.github.io/tag/software.html">software</a> <a href="https://standage.github.io/tag/rcgrep.html">rcgrep</a> <a href="https://standage.github.io/tag/frustration-driven-development.html">frustration driven development</a> </div>
	<div class="summary"><p>There are many wonderfully elegant and efficient tools for performing all sorts of exact and inexact searches on large collections of DNA sequences.
Experience has shown, however, that these tools are usually very rigid with respect to their assumptions about input data.
If input files are compressed in a certain way, or stored in a non-standard format, there is usually ...</p> <a class="btn primary xsmall" href="https://standage.github.io/super-simple-reverse-complement-aware-dna-sequence-search-with-rcgrep.html">more…</a>
	</div>
</div>	
<div class='article'>
	<h2><a href="https://standage.github.io/streaming-data-from-the-sra-with-fastq-dump.html">Streaming data from the SRA with fastq-dump</a></h2>
	<div class="well small">Permalink: <a class="more" href="https://standage.github.io/streaming-data-from-the-sra-with-fastq-dump.html">2017-01-24</a>
by <a class="url fn" href="https://standage.github.io/author/daniel-s-standage.html">Daniel S. Standage </a>
 in <a href="https://standage.github.io/category/blog.html">blog</a>
tags: <a href="https://standage.github.io/tag/sra.html">sra</a> <a href="https://standage.github.io/tag/streaming.html">streaming</a> <a href="https://standage.github.io/tag/ngs.html">ngs</a> </div>
	<div class="summary"><p>NCBI's <a href="https://www.ncbi.nlm.nih.gov/sra">Sequence Read Archive</a> is the go-to repository for published genome-scale sequence data sets.
Although there are a variety of ways to download sequence data from SRA, the <code>fastq-dump</code> command from the <a href="https://www.ncbi.nlm.nih.gov/Traces/sra/?view=software">SRA Toolkit</a> is the most convenient in my opinion.
In fact, with a few settings tweaks <code>fastq-dump</code> can stream data directly from the SRA into an analysis ...</p> <a class="btn primary xsmall" href="https://standage.github.io/streaming-data-from-the-sra-with-fastq-dump.html">more…</a>
	</div>
</div>	
<div class='article'>
	<h2><a href="https://standage.github.io/composing-generator-functions-in-python.html">Composing generator functions in Python</a></h2>
	<div class="well small">Permalink: <a class="more" href="https://standage.github.io/composing-generator-functions-in-python.html">2016-11-23</a>
by <a class="url fn" href="https://standage.github.io/author/daniel-s-standage.html">Daniel S. Standage </a>
 in <a href="https://standage.github.io/category/notebook.html">notebook</a>
tags: <a href="https://standage.github.io/tag/python.html">python</a> <a href="https://standage.github.io/tag/generators.html">generators</a> <a href="https://standage.github.io/tag/streaming.html">streaming</a> </div>
	<div class="summary"><p>In which I briefly motivate the utility of generator functions and
demonstrate that they can be nested to create a data processing stream.</p> <a class="btn primary xsmall" href="https://standage.github.io/composing-generator-functions-in-python.html">more…</a>
	</div>
</div>	
<div class='article'>
	<h2><a href="https://standage.github.io/thoughts-on-research-software-from-the-psrn-workshop.html">Thoughts on research software from the PSRN workshop</a></h2>
	<div class="well small">Permalink: <a class="more" href="https://standage.github.io/thoughts-on-research-software-from-the-psrn-workshop.html">2016-10-20</a>
by <a class="url fn" href="https://standage.github.io/author/daniel-s-standage.html">Daniel S. Standage </a>
 in <a href="https://standage.github.io/category/blog.html">blog</a>
tags: <a href="https://standage.github.io/tag/software.html">software</a> <a href="https://standage.github.io/tag/science.html">science</a> <a href="https://standage.github.io/tag/cyberinfrastructure.html">cyberinfrastructure</a> </div>
	<div class="summary"><p>In which I ramble on about experimental science, research software, and cyberinfrastructure engineering.</p> <a class="btn primary xsmall" href="https://standage.github.io/thoughts-on-research-software-from-the-psrn-workshop.html">more…</a>
	</div>
</div>	

<div class="pagination">
<ul>
		<li class="prev disabled"><a href="#">&larr; Previous</a></li>
		<li class="active"><a href="https://standage.github.io/author/daniel-s-standage.html">1</a></li>
		<li class=""><a href="https://standage.github.io/author/daniel-s-standage2.html">2</a></li>
		<li class=""><a href="https://standage.github.io/author/daniel-s-standage3.html">3</a></li>
		<li class=""><a href="https://standage.github.io/author/daniel-s-standage4.html">4</a></li>
		<li class=""><a href="https://standage.github.io/author/daniel-s-standage5.html">5</a></li>
		<li class="next"><a href="https://standage.github.io/author/daniel-s-standage2.html">Next &rarr;</a></li>
</ul>
</div>

		<footer>
		  <p>
			<a href="http://creativecommons.org/licenses/by/4.0/"><img src="https://standage.github.io/images/cc-by-88x31.png" alt="CC BY 4.0" /></a><br />
		    &copy; 2015-2018 Daniel S. Standage.<br />
			Unless otherwise noted, all content on this site is posted under a <a href="http://creativecommons.org/licenses/by/4.0/">CC BY license</a>.</p>
		  <p></p>
		  <p>This site is powered by <a href="http://getpelican.com/">Pelican</a>, its theme based on <a href="http://twitter.github.com/bootstrap/">Twitter Bootstrap</a>.</p>
		</footer>
	  </div>

	</div>
</body>
</html>