<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8">
	<title>Daniel Standage</title>
	<meta name="description" content="">
	<meta name="author" content="Daniel S. Standage">

	<!-- HTML5 shim, for IE6-8 support of HTML elements -->
	<!--[if lt IE 9]>
		<script src="https://standage.github.io/theme/html5.js"></script>
	<![endif]-->

	<!-- Styles -->
	<link href="https://standage.github.io/theme/bootstrap.min.css" rel="stylesheet">
	<link href="https://standage.github.io/theme/local.css" rel="stylesheet">
	<link href="https://standage.github.io/theme/pygments.css" rel="stylesheet">

	<!-- Feeds -->
	<link href="https://standage.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Daniel Standage Atom Feed" />


	<link href="https://standage.github.io/feed.xml" type="application/rss+xml" rel="alternate" title="Daniel Standage RSS Feed" />


</head>
<body>
	<div class="topbar">
	  <div class="topbar-inner">
		<div class="container-fluid">
		  <a class="brand" href="https://standage.github.io/">Daniel Standage</a>
			<ul class="nav">
						<li><a href="https://standage.github.io/pages/vitae.html">Vitae</a></li>
						<li><a href="https://standage.github.io/pages/teaching.html">Teaching</a></li>
						<li><a href="https://standage.github.io/pages/software-databases.html">Software & Databases</a></li>
				<li><a href="https://standage.github.io/notebook.html">Notebook</a></li>
			</ul>
			<p class="pull-right"><a href="https://standage.github.io/archives.html">[archives]</a> <a href="https://standage.github.io/categories.html">[categories]</a> <a href="https://standage.github.io/tags.html">[tags]</a></p>
		</div>
	  </div>
	</div>

	<div class="container-fluid">
	  <div class="sidebar">
		<div class="well">
			<img src="http://1.gravatar.com/avatar/fe326141c425326acf247aef9c475c03?size=180" alt="Daniel Standage" />
			<h3>Daniel S. Standage</h3>
			<p>
				205A Simon Hall<br />
				Indiana University<br />
				212 South Hawthorne Drive<br />
				Bloomington, IN 47405<br />
			</p>
			<h3>Affiliations</h3>
			<ul>
				<li><a href="http://brendelgroup.org/">Brendel Group</a></li>
				<li><a href="http://www.public.iastate.edu/~amytoth/Toth_lab/Home.html">Toth Lab</a></li>
				<li><a href="http://www.bio.indiana.edu/">Indiana Biology Department</a></li>
				<li><a href="http://www.bcb.iastate.edu/">Iowa State BCB Program</a></li>
			</ul>
			<div class="social">
			<h3>Social</h3>
			<ul>
				<li><a href="http://github.com/standage">GitHub</a></li>
				<li><a href="https://twitter.com/byuhobbes">Twitter</a></li>
				<li><a href="http://stackexchange.com/users/208980/daniel-standage?tab=accounts">StackExchange</a></li>
			</ul>
			</div>
		</div>
	  </div>
	  <div class="content">
	<div class='page'>
		<div class="page-header"><h1>Vitae</h1></div>
		<div><h2>Eductation</h2>
<table>
  <tr>
    <td>2010 — 2015</td>
    <td>
      Ph.D. in <strong>Bioinformatics and Computational Biology</strong><br />
      Iowa State University<br />
      In residence at Indiana University since July 2012
    </td>
  </tr>
  <tr>
    <td>2003 — 2010</td>
    <td>
      B.S. in <strong>Bioinformatics</strong> (with minors in <strong>Mathematics</strong> and <strong>Computer Science</strong>)<br />
      Brigham Young University<br />
      University Honors
    </td>
  </tr>
</table>

<h2>Scholarship</h2>
<h3>Publications</h3>
<ul>
<li>Duvick JP, <strong>Standage DS</strong>, Merchant N, Brendel VP (2015) xGDBvm: A Web GUI-driven workflow for annotating eukaryotic genomes in the cloud. <em>Manuscript in review</em>.</li>
<li><strong>Standage DS</strong>, Brendel VP (2012) ParsEval: parallel comparison and analysis of gene structure annotations. <em>BMC Bioinformatics</em>, <strong>13</strong>:187, <a href="http://dx.doi.org/10.1186/1471-2105-13-187">doi:10.1186/1471-2105-13-187</a>.</li>
<li>Doyle EL, Booher NJ, <strong>Standage DS</strong>, Voytas DF, Brendel VP, VanDyk JK, Bogdanove AJ (2012) TAL Effector-Nucleotide Targeter (TALE-NT) 2.0: Tools for TAL effector design and target prediction. <em>Nucleic Acids Research</em>, <strong>40</strong> (W1): W117-W122, <a href="http://dx.doi.org/10.1093/nar/gks608">doi:10.1093/nar/gks608</a>.</li>
</ul>
<h3>Oral Presentations</h3>
<ul>
<li><strong>Standage DS</strong>, Toth AL, Brendel VP. iLoci: scalable annotation for provisional genome assemblies. Presented at CSHL Biology and Genomics of Social Insects Meeting, May 2015. <em>Manuscript in preparation</em>.</li>
<li>Duvick JP, Denton JF, <strong>Standage DS</strong>, Merchant N, Brendel VP. xGDBvm: a virtual platform for annotating eukaryotic genomes. Invited talk given at PAG XXI, January 2013. <em>Manuscript in review</em>.</li>
</ul>
<h3>Posters</h3>
<ul>
<li><strong>Standage DS</strong>, Brendel VP, Toth AL. Genome sequence and annotation of the primitively social paper wasp <em>Polistes dominula</em>. Presented at Notre Dame Arthropod Genomics Conference, June 2013. <em>Manuscript submitted</em>.</li>
<li>Brendel VP, <strong>Standage DS</strong>. mRNAmarkup: quality control and annotation of de novo transcriptome assemblies. Presented at CSHL Genome Informatics meeting, November 2013. <em>Manuscript in preparation</em>.</li>
</ul>
<h2>Research Experience</h2>
<p><strong>Graduate Research Assistant</strong>, <small>Computational Genome Science Lab, Indiana University, 2012 — present</small></p>
<ul>
<li>Implement version control tools for genome annotations</li>
<li>Conduct comparative analysis of genome composition in social insects</li>
<li>Design new methods for organizing genomes to facilitate comparison and analysis between annotations, assemblies, and species</li>
<li>Investigate genome composition, differential expression, alternative splicing, and methylation in the paper wasp Polistes dominula</li>
</ul>
<p><strong>Graduate Research Assistant</strong>, <small>BCB Graduate Program, Iowa State University, 2010 — 2012</small></p>
<ul>
<li>Develop tools for comparing gene structure annotations</li>
<li>Implement parallel version of an NGS error correction program</li>
<li>Evaluate existing software tools for transposable element prediction</li>
</ul>
<p><strong>Undegraduate Research Assistant</strong>, <small>Plant Genetics Lab, Brigham Young University, 2009 — 2010</small></p>
<ul>
<li>Analyze expression of homoeologous genes in Gossypium (cotton)</li>
<li>Prototype sample submission system for on-site 454 sequencer</li>
<li>Develop tools for SNP analysis in polyploid organisms</li>
</ul>
<p><strong>Research Intern</strong>, <small>Department of Genetics, University of Georgia, 2008</small></p>
<ul>
<li>Investigate genes involved in biosynthesis of plant cell wall monosaccharides using phylogenetic profile analysis</li>
<li>Model gene regulatory networks in fungus Neurospora crassa</li>
</ul>
<p><strong>Undegraduate Research Assistant</strong>, <small>Center for Language Studies, Brigham Young University, 2007 — 2010</small></p>
<ul>
<li>Work with domain experts to produce responsive online learning materials</li>
<li>Research and prototype open, modular, and interoperable e-learning architectures</li>
</ul>
<h2>Professional Activities</h2>
<p><strong>Journal Review</strong></p>
<ul>
<li><strong>Science</strong>, May 2013</li>
<li><strong>Bioinformatics</strong>, April 2013</li>
</ul>
<p><strong>Open-source Software and Data</strong></p>
<ul>
<li><a href="http://standage.github.io/AEGeAn/">AEGeAn Toolkit</a>: primary developer</li>
<li><a href="http://brendelgroup.github.io/HymHub/">HymHub</a>: primary developer</li>
<li><a href="http://genometools.org/">GenomeTools</a> Library: contributor</li>
<li><a href="http://khmer.readthedocs.org/">khmer Library</a>: contributor</li>
</ul>
<p><strong>Community Training</strong></p>
<ul>
<li><a href="http://bioinformatics.msu.edu/ngs-summer-course-2014">2014 course on Analyzing NGS Data, Michigan State University</a>: invited instructor</li>
<li><a href="http://software-carpentry.org/">Software Carpentry</a>: certified instructor</li>
</ul>
<h2>Teaching</h2>
<p><strong>Computational Genome Science <small>(Fall 2011, Spring 2013, Spring 2014, Spring 2015, Spring 2016)</small></strong></p>
<p>Graduate course covering the full range of computational genomics analysis: quality control, read mapping, expression profiling, genome assembly, transcript assembly, and annotation.
The class provides brief exposure to relevant theory, but focuses primarily on installing and running software and (most importantly) critical analysis and biological interpretation of results.
I helped develop the course in 2011, and since 2013 I have been the primary instructor.</p>
<p><strong>Science of Biology, Honors <small>(Spring 2007, Spring 2008)</small></strong></p>
<p>As an undergraduate I was a teaching assistant for Prof. Craig Coleman's Biol 120H class.
This class provided a comprehensive introduction to the biological sciences, with modules for biochemistry, molecular biology, genetics, cell biology, and evolution.
The course focused on learning the scientific method with computer-based experiment simulations, and had a heavy emphasis on writing.</p></div>
	</div>
		<footer>
		  <p>
			<a href="http://creativecommons.org/licenses/by/4.0/"><img src="https://standage.github.io/images/cc-by-88x31.png" alt="CC BY 4.0" /></a><br />
		    &copy; 2015-2016 Daniel S. Standage.<br />
			Unless otherwise noted, all content on this site is posted under a <a href="http://creativecommons.org/licenses/by/4.0/">CC BY license</a>.</p>
		  <p></p>
		  <p>This site is powered by <a href="http://getpelican.com/">Pelican</a>, its theme based on <a href="http://twitter.github.com/bootstrap/">Twitter Bootstrap</a>.</p>
		</footer>
	  </div>

	</div>
</body>
</html>