Title: Vitae
navorder: 01

## Education

<table>
  <tr>
    <td>2016 — present</td>
    <td>
      Postdoctoral Scholar<br />
      Lab for Data Intensive Biology<br />
      University of California, Davis
    </td>
  </tr>
    <tr>
      <td>2010 — 2016</td>
      <td>
        Ph.D. in <strong>Bioinformatics and Computational Biology</strong><br />
        Iowa State University<br />
        In residence at Indiana University July 2012 — May 2016
      </td>
    </tr>
  <tr>
    <td>2003 — 2010</td>
    <td>
      B.S. in <strong>Bioinformatics</strong> (with minors in <strong>Mathematics</strong> and <strong>Computer Science</strong>)<br />
      Brigham Young University<br />
      University Honors
    </td>
  </tr>
</table>

## Scholarship

### Publications

- **Standage DS**, Crusoe MR, Joslin SEK, Kingsley NB, Murray KD, Neches R, Scott C, Shean R, Steinbiss S, Sydney C, Brown CT (2017). khmer release v2.1: software for biological sequence analysis. *In review*.
- **Standage DS**, Berens AJ, Glastad KM, Severin AJ, Brendel VP, Toth AL (2016) Genome, transcriptome, and methylome sequencing of a primitively eusocial wasp reveal a greatly reduced DNA methylation system in a social insect. *Molecular Ecology*, [doi:10.1111/mec.13578](http://dx.doi.org/10.1111/mec.13578).
- Duvick JP, **Standage DS**, Merchant N, Brendel VP (2016) xGDBvm: A Web GUI-driven workflow for annotating eukaryotic genomes in the cloud. *The Plant Cell*, Advance publication, [doi:10.1105/tpc.15.00933](http://dx.doi.org/10.1105/tpc.15.00933).
- **Standage DS**, Brendel VP (2012) ParsEval: parallel comparison and analysis of gene structure annotations. *BMC Bioinformatics*, **13**:187, [doi:10.1186/1471-2105-13-187](http://dx.doi.org/10.1186/1471-2105-13-187).
- Doyle EL, Booher NJ, **Standage DS**, Voytas DF, Brendel VP, VanDyk JK, Bogdanove AJ (2012) TAL Effector-Nucleotide Targeter (TALE-NT) 2.0: Tools for TAL effector design and target prediction. *Nucleic Acids Research*, **40** (W1): W117-W122, [doi:10.1093/nar/gks608](http://dx.doi.org/10.1093/nar/gks608).

### Oral Presentations

- **Standage DS**. Testing research software: an academic perspective on compromise. Presented at PyCon, May 2017.
- **Standage DS**, Brown CT, Hormozdiari F. Kevlar: reference-free variant discovery in human genomes and beyond. Presented at the Sequencing, Finishing, and Analysis of the Future Meeting, May 2017. *Manuscript in preparation*.
- **Standage DS**, Wilson GV. Managing research software projects. Presented at Lawrence Berkeley Laboratory LabTech, October 2016.
- **Standage DS**, Toth AL, Brendel VP. iLoci: scalable annotation for provisional genome assemblies. Presented at CSHL Biology and Genomics of Social Insects Meeting, May 2015. *Manuscript in preparation*.
- Duvick JP, Denton JF, **Standage DS**, Merchant N, Brendel VP. xGDBvm: a virtual platform for annotating eukaryotic genomes. Invited talk given at PAG XXI, January 2013.

### Posters

- **Standage DS**, Brendel VP, Toth AL. Genome sequence and annotation of the primitively social paper wasp *Polistes dominula*. Presented at Notre Dame Arthropod Genomics Conference, June 2013.
- Brendel VP, **Standage DS**. mRNAmarkup: quality control and annotation of de novo transcriptome assemblies. Presented at CSHL Genome Informatics meeting, November 2013.

## Research Experience

**Postdoctoral Scholar**, <small>Lab for Data Intensive Biology, UC Davis, 2016 – present</small>

- develop reference-free methods for variant discovery in large eukaryotic genomes
- lead development of [khmer DNA analysis software package](https://github.com/dib-lab/khmer)

**Graduate Research Assistant**, <small>Computational Genome Science Lab, Indiana University, 2012 — 2016</small>

- investigate comparative genomics of social insects
- investigate eukaryotic genome composition and organization
- develop tools for management, quality control and analysis of annotated genomes
- teach computational genome science course to biology and bioinformatics grad students

**Graduate Research Assistant**, <small>BCB Graduate Program, Iowa State University, 2010 — 2012</small>

- First year research rotations
    - develop tools for comparison of genome annotations
    - identify transposable elements in *Gossypium* (cotton)
    - implement multithreading support for short read error correction software
- Research assistantship
    - follow through with first research rotation to engineer and publish a tool for comparison of genome annotations (ParsEval)
    - assemble and annotate genome of the eusocial wasp *Polistes dominula*
    - assist with data management and migration of a large comparative plant genomics database (PlantGDB)
    - develop and participate in a course on NGS applications to genomics

**Undegraduate Research Technician**, <small>Plant Genetics Lab, Brigham Young University, 2009 — 2010</small>

- analyze homoeologous gene expression (genes duplicated by polyploidy) in cotton with high-throughput sequencing technology
- prototype an electronic sample submission system for on-site 454 sequencer
- develop software for SNP analysis in polyploid organisms
- assemble and map high-throughput sequence data (454 and Illumina)
- configure a genome browser for transcriptome visualization

**Research Intern**, <small>Department of Genetics, University of Georgia, 2008</small>

- model and simulate genetic networks in *Neurospora crassa*
- analyze biological clocks in fungi using basic experimental protocols (Northern blot, RNA extraction, gene electrophoresis, etc)
- introduction to systems biology and network modeling
- introduction to high-performance computing
- utilize phylogenetic profile analysis to identify genes involved with the biosynthesis of cell wall monosaccharides

**Undegraduate Research Assistant**, <small>Center for Language Studies, Brigham Young University, 2007 — 2010</small>

- develop applications and architectures to support development, delivery, and assessment of electronic learning resources and student performance
- focus on best practices such as separation of content, presentation, and logic
- work directly with subject matter experts and instructional designers to create instructional materials
- evangelize the topics of openness, modularity, and interoperability in the learning community
- served (briefly) as chair of the Architecture working group for LETSI (the International Federation for Learning, Education, and Training Systems Interoperability)

## Professional Activities

**Journal Review**

- **Science**, May 2013
- **Bioinformatics**, April 2013, October 2016

**Open-source Software and Data**

- [kevlar](http://kevlar.readthedocs.org): primary developer
- [khmer Library](http://khmer.readthedocs.org/): core developer
- [tag Library](https://github.com/standage/tag): primary developer
- [AEGeAn Toolkit](http://brendelgroup.github.io/AEGeAn): primary developer
- [*P. dominula* Genome Project](http://pdomgenomeproject.github.io/): lead
- [GenHub](http://standage.github.io/genhub): primary developer
- [GenomeTools Library](http://genometools.org/): contributor

**Community Training**

- [2017 Data Intensive Biology Summer Institute, UC Davis](http://ivory.idyll.org/dibsi/): instructor
- [2014 course on Analyzing NGS Data, Michigan State University](http://bioinformatics.msu.edu/ngs-summer-course-2014): invited instructor
- [Software Carpentry](http://software-carpentry.org/) certified instructor
- [Data Carpentry](http://www.datacarpentry.org/) certified instructor

## Technical Skills

- **Software development**: C/C++, Python, Java
- **Scripting, data processing**: Python, Perl, Bash
- **Data analysis, statistical computing**: Python, R
- **Pipelines, workflow management**: Bash, GNU Make
- **High-performance computing**: MPI programming, OpenMP programming, job scheduling and resource management, VM/cloud computing

## Teaching

**Computational Genome Science <small>(Fall 2011, Spring 2013, Spring 2014, Spring 2015, Spring 2016)</small>**

Graduate course covering the full range of computational genomics analysis: quality control, read mapping, expression profiling, genome assembly, transcript assembly, and annotation.
The class provides brief exposure to relevant theory, but focuses primarily on installing and running software and (most importantly) critical analysis and biological interpretation of results.
I helped develop the course in 2011, and since 2013 I have been the primary instructor.

**Science of Biology, Honors <small>(Spring 2007, Spring 2008)</small>**

As an undergraduate I was a teaching assistant for Prof. Craig Coleman's Biol 120H class.
This class provided a comprehensive introduction to the biological sciences, with modules for biochemistry, molecular biology, genetics, cell biology, and evolution.
The course focused on learning the scientific method with computer-based experiment simulations, and had a heavy emphasis on writing.
