Title: Software
navorder: 03

<div style="width: 42em;">

Except where specified (<em>in parentheses and italics</em>), I am the primary–and in most cases only–contributing developer to the following open source research software projects.

<h3><a href="http://kevlar.readthedocs.io/">kevlar Library</a></h3>

Alignment-free discovery of novel germline variants on simplex pedigrees.
Implemented as a library with command line and Python APIs, and distributed as a Python package.

<h3><a href="http://khmer.readthedocs.io/">khmer Library</a></h3>

In-memory <em>k</em>-mer counting, filtering, graph traversal, and more.
Implemented as a C++ library with Python bindings, and distributed as a Python package.
(<em>I'm a core contributor.</em>)

<h3><a href="http://tag.readthedocs.io/">tag Library</a></h3>

Genome annotation processing that supports graph-based representation of genome features and embraces a streaming processing paradigm.
Implemented as a library with a Python API (as well as a minimal command line interface), and distributed as a Python package.

<h3><a href="http://brendelgroup.github.io/AEGeAn">AEGeAn Toolkit</a></h3>

A collection of tools for analyzing and evaluating genome annotations, most notably <strong>ParsEval</strong> and <strong>LocusPocus</strong>.
Implemented as a library with a C API.

<h3><a href="http://standage.github.io/genhub">GenHub</a></h3>

A companion to the AEGeAn Toolkit, providing a more comprehensive pipeline around <strong>LocusPocus</strong> for data retrieval, pre-processing, post-processing, and computing statistics on iLoci.
Implemented in Python.

<h3><a href="http://genometools.org/">GenomeTools Library</a></h3>

A comprehensive toolkit for processing, transforming, and analyzing genome annotation data.
Implemented as a library with a C API.
(<em>I'm a contributor.</em>)

</div>
